{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Datacamp - RNA-Seq with Bioconductor in R\n",
    "RNA-Seq is an exciting next-generation sequencing method used for identifying genes and pathways underlying particular diseases or conditions. As high-throughput sequencing becomes more affordable and accessible to a wider community of researchers, the knowledge to analyze this data is becoming an increasingly valuable skill. Join us in learning about the RNA-Seq workflow and discovering how to identify which genes and biological processes may be important for your condition of interest! We will start the course with a brief overview of the RNA-Seq workflow with an emphasis on differential expression (DE) analysis. Starting with the counts for each gene, the course will cover how to prepare data for DE analysis, assess the quality of the count data, and identify outliers and detect major sources of variation in the data. The DESeq2 R package will be used to model the count data using a negative binomial model and test for differentially expressed genes. Visualization of the results with heatmaps and volcano plots will be performed and the significant differentially expressed genes will be identified and saved."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction to RNA-Seq theory and workflow\n",
    "In this chapter we explore what we can do with RNA-Seq data and why it is exciting. We learn about the different steps and considerations involved in an RNA-Seq workflow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA-Seq Packages\n",
    "We will be using DESeq2 for performing the differential expression analysis and additional R packages for data wrangling and plotting. Before we can run any analyses, we need to load the following packages `DESeq2`, `RColorBrewer`, `pheatmap`, and `tidyverse`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)\n",
      "\n",
      "Installing package(s) 'DESeq2'\n",
      "\n",
      "also installing the dependencies 'annotate', 'genefilter', 'locfit', 'geneplotter', 'RcppArmadillo'\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'annotate' successfully unpacked and MD5 sums checked\n",
      "package 'genefilter' successfully unpacked and MD5 sums checked\n",
      "package 'locfit' successfully unpacked and MD5 sums checked\n",
      "package 'geneplotter' successfully unpacked and MD5 sums checked\n",
      "package 'RcppArmadillo' successfully unpacked and MD5 sums checked\n",
      "package 'DESeq2' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\recer\\AppData\\Local\\Temp\\Rtmp2L9qwm\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'backports', 'brio', 'broom', 'cli', 'cpp11', 'crayon', 'DBI',\n",
      "  'digest', 'dtplyr', 'fansi', 'fs', 'generics', 'glmnet', 'glue', 'jsonlite',\n",
      "  'knitr', 'languageserver', 'magrittr', 'mvtnorm', 'openssl', 'permute',\n",
      "  'pillar', 'pkgload', 'raster', 'Rcpp', 'readr', 'remotes', 'repr', 'rex',\n",
      "  'rJava', 'rjson', 'rlang', 'rsconnect', 'rvest', 'sp', 'stringi', 'testthat',\n",
      "  'tibble', 'tinytex', 'tzdb', 'uuid', 'vroom', 'withr', 'xfun', 'xgboost',\n",
      "  'xml2', 'yaml', 'foreign', 'later', 'rpart'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'RColorBrewer'\"\n",
      "Old packages: 'backports', 'brio', 'broom', 'cli', 'cpp11', 'crayon', 'DBI',\n",
      "  'digest', 'dtplyr', 'fansi', 'fs', 'generics', 'glmnet', 'glue', 'jsonlite',\n",
      "  'knitr', 'languageserver', 'magrittr', 'mvtnorm', 'openssl', 'permute',\n",
      "  'pillar', 'pkgload', 'raster', 'Rcpp', 'readr', 'remotes', 'repr', 'rex',\n",
      "  'rJava', 'rjson', 'rlang', 'rsconnect', 'rvest', 'sp', 'stringi', 'testthat',\n",
      "  'tibble', 'tinytex', 'tzdb', 'uuid', 'vroom', 'withr', 'xfun', 'xgboost',\n",
      "  'xml2', 'yaml', 'foreign', 'later', 'rpart'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)\n",
      "\n",
      "Installing package(s) 'pheatmap'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'pheatmap' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\recer\\AppData\\Local\\Temp\\Rtmp2L9qwm\\downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'backports', 'brio', 'broom', 'cli', 'cpp11', 'crayon', 'DBI',\n",
      "  'digest', 'dtplyr', 'fansi', 'fs', 'generics', 'glmnet', 'glue', 'jsonlite',\n",
      "  'knitr', 'languageserver', 'magrittr', 'mvtnorm', 'openssl', 'permute',\n",
      "  'pillar', 'pkgload', 'raster', 'Rcpp', 'readr', 'remotes', 'repr', 'rex',\n",
      "  'rJava', 'rjson', 'rlang', 'rsconnect', 'rvest', 'sp', 'stringi', 'testthat',\n",
      "  'tibble', 'tinytex', 'tzdb', 'uuid', 'vroom', 'withr', 'xfun', 'xgboost',\n",
      "  'xml2', 'yaml', 'foreign', 'later', 'rpart'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Install library for DESeq2\n",
    "BiocManager::install(\"DESeq2\")\n",
    "\n",
    "# Install library for RColorBrewer\n",
    "BiocManager::install(\"RColorBrewer\")\n",
    "\n",
    "# Install library for pheatmap\n",
    "BiocManager::install(\"pheatmap\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load library for DESeq2\n",
    "library(\"DESeq2\")\n",
    "\n",
    "# Load library for RColorBrewer\n",
    "library(\"RColorBrewer\")\n",
    "\n",
    "# Load library for pheatmap\n",
    "library(\"pheatmap\")\n",
    "\n",
    "# Load library for tidyverse\n",
    "library(\"tidyverse\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the raw count matrix\n",
    "Let's explore the contents of the raw count matrix. The dataset smoc2_rawcounts has already been loaded for you using the `read.csv()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read CSV read file\n",
    "smoc2_rawcounts <- read.csv('data/smoc2_rawcounts.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>smoc2_fibrosis1</th><th scope=col>smoc2_fibrosis4</th><th scope=col>smoc2_normal1</th><th scope=col>smoc2_normal3</th><th scope=col>smoc2_fibrosis3</th><th scope=col>smoc2_normal4</th><th scope=col>smoc2_fibrosis2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSMUSG00000102693</td><td> 0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSMUSG00000064842</td><td> 0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSMUSG00000051951</td><td>72</td><td>30</td><td>0</td><td>3</td><td>36</td><td>1</td><td>51</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSMUSG00000102851</td><td> 0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSMUSG00000103377</td><td> 0</td><td> 0</td><td>1</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSMUSG00000104017</td><td> 0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & X & smoc2\\_fibrosis1 & smoc2\\_fibrosis4 & smoc2\\_normal1 & smoc2\\_normal3 & smoc2\\_fibrosis3 & smoc2\\_normal4 & smoc2\\_fibrosis2\\\\\n",
       "  & <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENSMUSG00000102693 &  0 &  0 & 0 & 0 &  0 & 0 &  0\\\\\n",
       "\t2 & ENSMUSG00000064842 &  0 &  0 & 0 & 0 &  0 & 0 &  0\\\\\n",
       "\t3 & ENSMUSG00000051951 & 72 & 30 & 0 & 3 & 36 & 1 & 51\\\\\n",
       "\t4 & ENSMUSG00000102851 &  0 &  0 & 0 & 0 &  0 & 0 &  0\\\\\n",
       "\t5 & ENSMUSG00000103377 &  0 &  0 & 1 & 0 &  0 & 0 &  0\\\\\n",
       "\t6 & ENSMUSG00000104017 &  0 &  0 & 0 & 0 &  0 & 0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | X &lt;chr&gt; | smoc2_fibrosis1 &lt;int&gt; | smoc2_fibrosis4 &lt;int&gt; | smoc2_normal1 &lt;int&gt; | smoc2_normal3 &lt;int&gt; | smoc2_fibrosis3 &lt;int&gt; | smoc2_normal4 &lt;int&gt; | smoc2_fibrosis2 &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSMUSG00000102693 |  0 |  0 | 0 | 0 |  0 | 0 |  0 |\n",
       "| 2 | ENSMUSG00000064842 |  0 |  0 | 0 | 0 |  0 | 0 |  0 |\n",
       "| 3 | ENSMUSG00000051951 | 72 | 30 | 0 | 3 | 36 | 1 | 51 |\n",
       "| 4 | ENSMUSG00000102851 |  0 |  0 | 0 | 0 |  0 | 0 |  0 |\n",
       "| 5 | ENSMUSG00000103377 |  0 |  0 | 1 | 0 |  0 | 0 |  0 |\n",
       "| 6 | ENSMUSG00000104017 |  0 |  0 | 0 | 0 |  0 | 0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "  X                  smoc2_fibrosis1 smoc2_fibrosis4 smoc2_normal1\n",
       "1 ENSMUSG00000102693  0               0              0            \n",
       "2 ENSMUSG00000064842  0               0              0            \n",
       "3 ENSMUSG00000051951 72              30              0            \n",
       "4 ENSMUSG00000102851  0               0              0            \n",
       "5 ENSMUSG00000103377  0               0              1            \n",
       "6 ENSMUSG00000104017  0               0              0            \n",
       "  smoc2_normal3 smoc2_fibrosis3 smoc2_normal4 smoc2_fibrosis2\n",
       "1 0              0              0              0             \n",
       "2 0              0              0              0             \n",
       "3 3             36              1             51             \n",
       "4 0              0              0              0             \n",
       "5 0              0              0              0             \n",
       "6 0              0              0              0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t47729 obs. of  8 variables:\n",
      " $ X              : chr  \"ENSMUSG00000102693\" \"ENSMUSG00000064842\" \"ENSMUSG00000051951\" \"ENSMUSG00000102851\" ...\n",
      " $ smoc2_fibrosis1: int  0 0 72 0 0 0 0 0 0 1 ...\n",
      " $ smoc2_fibrosis4: int  0 0 30 0 0 0 0 0 0 1 ...\n",
      " $ smoc2_normal1  : int  0 0 0 0 1 0 0 0 0 1 ...\n",
      " $ smoc2_normal3  : int  0 0 3 0 0 0 0 0 0 0 ...\n",
      " $ smoc2_fibrosis3: int  0 0 36 0 0 0 0 0 0 1 ...\n",
      " $ smoc2_normal4  : int  0 0 1 0 0 0 0 0 0 0 ...\n",
      " $ smoc2_fibrosis2: int  0 0 51 0 0 0 0 0 0 1 ...\n"
     ]
    }
   ],
   "source": [
    "# Explore the first six observations of smoc2_rawcounts\n",
    "head(smoc2_rawcounts)\n",
    "\n",
    "# Explore the structure of smoc2_rawcounts\n",
    "str(smoc2_rawcounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create genotype vector\n",
    "genotype <- c(\"smoc2_oe\", \"smoc2_oe\", \"smoc2_oe\", \"smoc2_oe\", \"smoc2_oe\", \"smoc2_oe\", \"smoc2_oe\")\n",
    "\n",
    "# Create condition vector\n",
    "condition <- c(\"fibrosis\", \"fibrosis\", \"fibrosis\", \"fibrosis\", \"normal\", \"normal\", \"normal\")\n",
    "\n",
    "# Create data frame\n",
    "smoc2_metadata <- data.frame(genotype, condition)\n",
    "\n",
    "# Assign the row names of the data frame\n",
    "rownames(smoc2_metadata) <- c(\"smoc2_fibrosis1\", \"smoc2_fibrosis2\", \"smoc2_fibrosis3\", \"smoc2_fibrosis4\", \"smoc2_normal1\", \"smoc2_normal3\", \"smoc2_normal4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 7 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>genotype</th><th scope=col>condition</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>smoc2_fibrosis1</th><td>smoc2_oe</td><td>fibrosis</td></tr>\n",
       "\t<tr><th scope=row>smoc2_fibrosis2</th><td>smoc2_oe</td><td>fibrosis</td></tr>\n",
       "\t<tr><th scope=row>smoc2_fibrosis3</th><td>smoc2_oe</td><td>fibrosis</td></tr>\n",
       "\t<tr><th scope=row>smoc2_fibrosis4</th><td>smoc2_oe</td><td>fibrosis</td></tr>\n",
       "\t<tr><th scope=row>smoc2_normal1</th><td>smoc2_oe</td><td>normal  </td></tr>\n",
       "\t<tr><th scope=row>smoc2_normal3</th><td>smoc2_oe</td><td>normal  </td></tr>\n",
       "\t<tr><th scope=row>smoc2_normal4</th><td>smoc2_oe</td><td>normal  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & genotype & condition\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tsmoc2\\_fibrosis1 & smoc2\\_oe & fibrosis\\\\\n",
       "\tsmoc2\\_fibrosis2 & smoc2\\_oe & fibrosis\\\\\n",
       "\tsmoc2\\_fibrosis3 & smoc2\\_oe & fibrosis\\\\\n",
       "\tsmoc2\\_fibrosis4 & smoc2\\_oe & fibrosis\\\\\n",
       "\tsmoc2\\_normal1 & smoc2\\_oe & normal  \\\\\n",
       "\tsmoc2\\_normal3 & smoc2\\_oe & normal  \\\\\n",
       "\tsmoc2\\_normal4 & smoc2\\_oe & normal  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7 × 2\n",
       "\n",
       "| <!--/--> | genotype &lt;chr&gt; | condition &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| smoc2_fibrosis1 | smoc2_oe | fibrosis |\n",
       "| smoc2_fibrosis2 | smoc2_oe | fibrosis |\n",
       "| smoc2_fibrosis3 | smoc2_oe | fibrosis |\n",
       "| smoc2_fibrosis4 | smoc2_oe | fibrosis |\n",
       "| smoc2_normal1 | smoc2_oe | normal   |\n",
       "| smoc2_normal3 | smoc2_oe | normal   |\n",
       "| smoc2_normal4 | smoc2_oe | normal   |\n",
       "\n"
      ],
      "text/plain": [
       "                genotype condition\n",
       "smoc2_fibrosis1 smoc2_oe fibrosis \n",
       "smoc2_fibrosis2 smoc2_oe fibrosis \n",
       "smoc2_fibrosis3 smoc2_oe fibrosis \n",
       "smoc2_fibrosis4 smoc2_oe fibrosis \n",
       "smoc2_normal1   smoc2_oe normal   \n",
       "smoc2_normal3   smoc2_oe normal   \n",
       "smoc2_normal4   smoc2_oe normal   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "smoc2_metadata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "starting httpd help server ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vignette(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Matching metadata and counts data\n",
    "To perform any analysis with DESeq2, we need to create a DESeq2 object by providing the raw counts, metadata, and design formula. To do this, we need to read in the raw counts data and associated metadata we created previously, make sure the sample names are in the same order in both datasets, then create a DESeq2 object to use for differential expression analysis. We will use the design formula `~ condition` to test for differential expression between conditions (normal and fibrosis).\n",
    "\n",
    "The `DESeq2` and `dplyr` libraries have been loaded for you, and the smoc2_rawcounts and smoc2_metadata files have been read in.\n",
    "\n",
    "- Use the `match()` function to return the indices for how to reorder the columns of the counts data to match the order of the row names of the metadata. Assign the result to reorder_idx.\n",
    "- Remember that when using the match() function, the first vector is in the order desired, while the second vector is the one to reorder.\n",
    "- To create the DESeq2 object, we need to specify our reordered counts (reordered_smoc2_rawcounts) as the countData and our metadata (smoc2_metadata) as the colData.\n",
    "\n",
    "- Reorder the columns of the count data with reorder_idx such that the column names match the order of the row names in the metadata.\n",
    "\n",
    "- Create a DESeq2 object, dds_smoc2 using the `DESeqDataSetFromMatrix()` function using the metadata and reordered counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>2</li><li>8</li><li>6</li><li>3</li><li>4</li><li>5</li><li>7</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 2\n",
       "\\item 8\n",
       "\\item 6\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\item 5\n",
       "\\item 7\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 2\n",
       "2. 8\n",
       "3. 6\n",
       "4. 3\n",
       "5. 4\n",
       "6. 5\n",
       "7. 7\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 2 8 6 3 4 5 7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use the match() function to reorder the columns of the raw counts\n",
    "reorder_idx <- match(rownames(smoc2_metadata), colnames(smoc2_rawcounts))\n",
    "reorder_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "\"some variables in design formula are characters, converting to factors\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: DESeqDataSet \n",
       "dim: 47729 7 \n",
       "metadata(1): version\n",
       "assays(1): counts\n",
       "rownames: NULL\n",
       "rowData names(0):\n",
       "colnames(7): smoc2_fibrosis1 smoc2_fibrosis2 ... smoc2_normal3\n",
       "  smoc2_normal4\n",
       "colData names(2): genotype condition"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Reorder the columns of the count data\n",
    "reordered_smoc2_rawcounts <- smoc2_rawcounts[ , reorder_idx]\n",
    "\n",
    "# Create a DESeq2 object\n",
    "dds_smoc2 <- DESeqDataSetFromMatrix(countData = reordered_smoc2_rawcounts,\n",
    "                              colData = smoc2_metadata,\n",
    "                              design = ~ condition)\n",
    "dds_smoc2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalizing counts with DESeq2\n",
    "We have created the DESeq2 object and now wish to perform quality control on our samples. Therefore, we need to generate the normalized counts (normalized for library size, which is the total number of gene counts per sample, while accounting for library composition). To obtain the normalized counts, use the DESeq2 object and generate the normalized counts matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>smoc2_fibrosis1</dt><dd>1.43198320905838</dd><dt>smoc2_fibrosis2</dt><dd>1.11896424564986</dd><dt>smoc2_fibrosis3</dt><dd>1.2480023540238</dd><dt>smoc2_fibrosis4</dt><dd>1.08267987463881</dd><dt>smoc2_normal1</dt><dd>0.710648210882817</dd><dt>smoc2_normal3</dt><dd>0.798973447160559</dd><dt>smoc2_normal4</dt><dd>0.848242621232058</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[smoc2\\textbackslash{}\\_fibrosis1] 1.43198320905838\n",
       "\\item[smoc2\\textbackslash{}\\_fibrosis2] 1.11896424564986\n",
       "\\item[smoc2\\textbackslash{}\\_fibrosis3] 1.2480023540238\n",
       "\\item[smoc2\\textbackslash{}\\_fibrosis4] 1.08267987463881\n",
       "\\item[smoc2\\textbackslash{}\\_normal1] 0.710648210882817\n",
       "\\item[smoc2\\textbackslash{}\\_normal3] 0.798973447160559\n",
       "\\item[smoc2\\textbackslash{}\\_normal4] 0.848242621232058\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "smoc2_fibrosis1\n",
       ":   1.43198320905838smoc2_fibrosis2\n",
       ":   1.11896424564986smoc2_fibrosis3\n",
       ":   1.2480023540238smoc2_fibrosis4\n",
       ":   1.08267987463881smoc2_normal1\n",
       ":   0.710648210882817smoc2_normal3\n",
       ":   0.798973447160559smoc2_normal4\n",
       ":   0.848242621232058\n",
       "\n"
      ],
      "text/plain": [
       "smoc2_fibrosis1 smoc2_fibrosis2 smoc2_fibrosis3 smoc2_fibrosis4   smoc2_normal1 \n",
       "      1.4319832       1.1189642       1.2480024       1.0826799       0.7106482 \n",
       "  smoc2_normal3   smoc2_normal4 \n",
       "      0.7989734       0.8482426 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 47729 × 7 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>smoc2_fibrosis1</th><th scope=col>smoc2_fibrosis2</th><th scope=col>smoc2_fibrosis3</th><th scope=col>smoc2_fibrosis4</th><th scope=col>smoc2_normal1</th><th scope=col>smoc2_normal3</th><th scope=col>smoc2_normal4</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td> 50.2799192</td><td> 45.5778638</td><td> 28.8460994</td><td> 27.7090216</td><td>  0.000000</td><td>  3.754818</td><td>  1.178908</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  1.407166</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.6983322</td><td>  0.8936836</td><td>  0.8012805</td><td>  0.9236341</td><td>  1.407166</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.9236341</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  6.2849899</td><td>  2.6810508</td><td>  5.6089638</td><td>  1.8472681</td><td>  1.407166</td><td>  1.251606</td><td>  1.178908</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  3.4916611</td><td>  2.6810508</td><td>  7.2115249</td><td>  3.6945362</td><td> 15.478826</td><td> 21.277303</td><td> 30.651608</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>339.3894544</td><td>279.7229681</td><td>342.1467905</td><td>237.3739515</td><td>368.677492</td><td>445.571754</td><td>418.512335</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.8936836</td><td>  0.0000000</td><td>  0.0000000</td><td>  1.407166</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  4.8883255</td><td>  4.4684180</td><td>  4.0064027</td><td>  1.8472681</td><td> 16.885992</td><td> 11.264454</td><td>  3.536724</td></tr>\n",
       "\t<tr><td>  1.3966644</td><td>  0.8936836</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  1.251606</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  1.3966644</td><td>  0.8936836</td><td>  2.4038416</td><td>  3.6945362</td><td>  4.221498</td><td>  0.000000</td><td>  2.357816</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.8936836</td><td>  0.0000000</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  1.178908</td></tr>\n",
       "\t<tr><td>  4.8883255</td><td> 16.0863049</td><td>  8.0128054</td><td>  8.3127065</td><td>  2.814332</td><td>  7.509636</td><td> 12.967988</td></tr>\n",
       "\t<tr><td>  2.0949966</td><td>  4.4684180</td><td>  0.8012805</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>  0.0000000</td><td>  0.8936836</td><td>  0.8012805</td><td>  0.0000000</td><td>  0.000000</td><td>  0.000000</td><td>  0.000000</td></tr>\n",
       "\t<tr><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   1.3966644</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.6983322</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.6983322</td><td>   3.574734</td><td>   0.8012805</td><td>   7.3890724</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>4110.3833919</td><td>3521.113400</td><td>3840.5376276</td><td>4067.6843665</td><td>2544.1561</td><td>2055.1371</td><td>1661.081352</td></tr>\n",
       "\t<tr><td> 569.8390839</td><td> 655.070082</td><td> 655.4474816</td><td> 549.5622611</td><td> 716.2475</td><td> 699.6478</td><td> 811.088694</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.8012805</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.9236341</td><td>   0.0000</td><td>   0.0000</td><td>   1.178908</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td>   0.0000000</td><td>   0.000000</td><td>   0.0000000</td><td>   0.0000000</td><td>   0.0000</td><td>   0.0000</td><td>   0.000000</td></tr>\n",
       "\t<tr><td> 975.5700983</td><td>1026.842461</td><td>1318.9077686</td><td> 621.6057172</td><td> 270.1759</td><td> 287.8694</td><td> 278.222285</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 47729 × 7 of type dbl\n",
       "\\begin{tabular}{lllllll}\n",
       " smoc2\\_fibrosis1 & smoc2\\_fibrosis2 & smoc2\\_fibrosis3 & smoc2\\_fibrosis4 & smoc2\\_normal1 & smoc2\\_normal3 & smoc2\\_normal4\\\\\n",
       "\\hline\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t  50.2799192 &  45.5778638 &  28.8460994 &  27.7090216 &   0.000000 &   3.754818 &   1.178908\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   1.407166 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.6983322 &   0.8936836 &   0.8012805 &   0.9236341 &   1.407166 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.9236341 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   6.2849899 &   2.6810508 &   5.6089638 &   1.8472681 &   1.407166 &   1.251606 &   1.178908\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   3.4916611 &   2.6810508 &   7.2115249 &   3.6945362 &  15.478826 &  21.277303 &  30.651608\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t 339.3894544 & 279.7229681 & 342.1467905 & 237.3739515 & 368.677492 & 445.571754 & 418.512335\\\\\n",
       "\t   0.0000000 &   0.8936836 &   0.0000000 &   0.0000000 &   1.407166 &   0.000000 &   0.000000\\\\\n",
       "\t   4.8883255 &   4.4684180 &   4.0064027 &   1.8472681 &  16.885992 &  11.264454 &   3.536724\\\\\n",
       "\t   1.3966644 &   0.8936836 &   0.0000000 &   0.0000000 &   0.000000 &   1.251606 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.0000000 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   1.3966644 &   0.8936836 &   2.4038416 &   3.6945362 &   4.221498 &   0.000000 &   2.357816\\\\\n",
       "\t   0.0000000 &   0.8936836 &   0.0000000 &   0.0000000 &   0.000000 &   0.000000 &   1.178908\\\\\n",
       "\t   4.8883255 &  16.0863049 &   8.0128054 &   8.3127065 &   2.814332 &   7.509636 &  12.967988\\\\\n",
       "\t   2.0949966 &   4.4684180 &   0.8012805 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t   0.0000000 &   0.8936836 &   0.8012805 &   0.0000000 &   0.000000 &   0.000000 &   0.000000\\\\\n",
       "\t ... & ... & ... & ... & ... & ... & ...\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    1.3966644 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.6983322 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.6983322 &    3.574734 &    0.8012805 &    7.3890724 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t 4110.3833919 & 3521.113400 & 3840.5376276 & 4067.6843665 & 2544.1561 & 2055.1371 & 1661.081352\\\\\n",
       "\t  569.8390839 &  655.070082 &  655.4474816 &  549.5622611 &  716.2475 &  699.6478 &  811.088694\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.8012805 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.9236341 &    0.0000 &    0.0000 &    1.178908\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t    0.0000000 &    0.000000 &    0.0000000 &    0.0000000 &    0.0000 &    0.0000 &    0.000000\\\\\n",
       "\t  975.5700983 & 1026.842461 & 1318.9077686 &  621.6057172 &  270.1759 &  287.8694 &  278.222285\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 47729 × 7 of type dbl\n",
       "\n",
       "| smoc2_fibrosis1 | smoc2_fibrosis2 | smoc2_fibrosis3 | smoc2_fibrosis4 | smoc2_normal1 | smoc2_normal3 | smoc2_normal4 |\n",
       "|---|---|---|---|---|---|---|\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|  50.2799192 |  45.5778638 |  28.8460994 |  27.7090216 |   0.000000 |   3.754818 |   1.178908 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   1.407166 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.6983322 |   0.8936836 |   0.8012805 |   0.9236341 |   1.407166 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.9236341 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   6.2849899 |   2.6810508 |   5.6089638 |   1.8472681 |   1.407166 |   1.251606 |   1.178908 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   3.4916611 |   2.6810508 |   7.2115249 |   3.6945362 |  15.478826 |  21.277303 |  30.651608 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "| 339.3894544 | 279.7229681 | 342.1467905 | 237.3739515 | 368.677492 | 445.571754 | 418.512335 |\n",
       "|   0.0000000 |   0.8936836 |   0.0000000 |   0.0000000 |   1.407166 |   0.000000 |   0.000000 |\n",
       "|   4.8883255 |   4.4684180 |   4.0064027 |   1.8472681 |  16.885992 |  11.264454 |   3.536724 |\n",
       "|   1.3966644 |   0.8936836 |   0.0000000 |   0.0000000 |   0.000000 |   1.251606 |   0.000000 |\n",
       "|   0.0000000 |   0.0000000 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   1.3966644 |   0.8936836 |   2.4038416 |   3.6945362 |   4.221498 |   0.000000 |   2.357816 |\n",
       "|   0.0000000 |   0.8936836 |   0.0000000 |   0.0000000 |   0.000000 |   0.000000 |   1.178908 |\n",
       "|   4.8883255 |  16.0863049 |   8.0128054 |   8.3127065 |   2.814332 |   7.509636 |  12.967988 |\n",
       "|   2.0949966 |   4.4684180 |   0.8012805 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "|   0.0000000 |   0.8936836 |   0.8012805 |   0.0000000 |   0.000000 |   0.000000 |   0.000000 |\n",
       "| ... | ... | ... | ... | ... | ... | ... |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    1.3966644 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.6983322 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.6983322 |    3.574734 |    0.8012805 |    7.3890724 |    0.0000 |    0.0000 |    0.000000 |\n",
       "| 4110.3833919 | 3521.113400 | 3840.5376276 | 4067.6843665 | 2544.1561 | 2055.1371 | 1661.081352 |\n",
       "|  569.8390839 |  655.070082 |  655.4474816 |  549.5622611 |  716.2475 |  699.6478 |  811.088694 |\n",
       "|    0.0000000 |    0.000000 |    0.8012805 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.9236341 |    0.0000 |    0.0000 |    1.178908 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|    0.0000000 |    0.000000 |    0.0000000 |    0.0000000 |    0.0000 |    0.0000 |    0.000000 |\n",
       "|  975.5700983 | 1026.842461 | 1318.9077686 |  621.6057172 |  270.1759 |  287.8694 |  278.222285 |\n",
       "\n"
      ],
      "text/plain": [
       "      smoc2_fibrosis1 smoc2_fibrosis2 smoc2_fibrosis3 smoc2_fibrosis4\n",
       " [1,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [2,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [3,]  50.2799192      45.5778638      28.8460994      27.7090216    \n",
       " [4,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [5,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [6,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [7,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [8,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       " [9,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[10,]   0.6983322       0.8936836       0.8012805       0.9236341    \n",
       "[11,]   0.0000000       0.0000000       0.0000000       0.9236341    \n",
       "[12,]   6.2849899       2.6810508       5.6089638       1.8472681    \n",
       "[13,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[14,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[15,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[16,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[17,]   3.4916611       2.6810508       7.2115249       3.6945362    \n",
       "[18,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[19,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[20,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[21,] 339.3894544     279.7229681     342.1467905     237.3739515    \n",
       "[22,]   0.0000000       0.8936836       0.0000000       0.0000000    \n",
       "[23,]   4.8883255       4.4684180       4.0064027       1.8472681    \n",
       "[24,]   1.3966644       0.8936836       0.0000000       0.0000000    \n",
       "[25,]   0.0000000       0.0000000       0.0000000       0.0000000    \n",
       "[26,]   1.3966644       0.8936836       2.4038416       3.6945362    \n",
       "[27,]   0.0000000       0.8936836       0.0000000       0.0000000    \n",
       "[28,]   4.8883255      16.0863049       8.0128054       8.3127065    \n",
       "[29,]   2.0949966       4.4684180       0.8012805       0.0000000    \n",
       "[30,]   0.0000000       0.8936836       0.8012805       0.0000000    \n",
       "[31,] ...             ...             ...             ...            \n",
       "[32,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[33,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[34,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[35,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[36,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[37,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[38,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[39,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[40,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[41,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[42,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[43,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[44,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[45,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[46,]    1.3966644       0.000000        0.0000000       0.0000000   \n",
       "[47,]    0.6983322       0.000000        0.0000000       0.0000000   \n",
       "[48,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[49,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[50,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[51,]    0.6983322       3.574734        0.8012805       7.3890724   \n",
       "[52,] 4110.3833919    3521.113400     3840.5376276    4067.6843665   \n",
       "[53,]  569.8390839     655.070082      655.4474816     549.5622611   \n",
       "[54,]    0.0000000       0.000000        0.8012805       0.0000000   \n",
       "[55,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[56,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[57,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[58,]    0.0000000       0.000000        0.0000000       0.9236341   \n",
       "[59,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[60,]    0.0000000       0.000000        0.0000000       0.0000000   \n",
       "[61,]  975.5700983    1026.842461     1318.9077686     621.6057172   \n",
       "      smoc2_normal1 smoc2_normal3 smoc2_normal4\n",
       " [1,]   0.000000      0.000000      0.000000   \n",
       " [2,]   0.000000      0.000000      0.000000   \n",
       " [3,]   0.000000      3.754818      1.178908   \n",
       " [4,]   0.000000      0.000000      0.000000   \n",
       " [5,]   1.407166      0.000000      0.000000   \n",
       " [6,]   0.000000      0.000000      0.000000   \n",
       " [7,]   0.000000      0.000000      0.000000   \n",
       " [8,]   0.000000      0.000000      0.000000   \n",
       " [9,]   0.000000      0.000000      0.000000   \n",
       "[10,]   1.407166      0.000000      0.000000   \n",
       "[11,]   0.000000      0.000000      0.000000   \n",
       "[12,]   1.407166      1.251606      1.178908   \n",
       "[13,]   0.000000      0.000000      0.000000   \n",
       "[14,]   0.000000      0.000000      0.000000   \n",
       "[15,]   0.000000      0.000000      0.000000   \n",
       "[16,]   0.000000      0.000000      0.000000   \n",
       "[17,]  15.478826     21.277303     30.651608   \n",
       "[18,]   0.000000      0.000000      0.000000   \n",
       "[19,]   0.000000      0.000000      0.000000   \n",
       "[20,]   0.000000      0.000000      0.000000   \n",
       "[21,] 368.677492    445.571754    418.512335   \n",
       "[22,]   1.407166      0.000000      0.000000   \n",
       "[23,]  16.885992     11.264454      3.536724   \n",
       "[24,]   0.000000      1.251606      0.000000   \n",
       "[25,]   0.000000      0.000000      0.000000   \n",
       "[26,]   4.221498      0.000000      2.357816   \n",
       "[27,]   0.000000      0.000000      1.178908   \n",
       "[28,]   2.814332      7.509636     12.967988   \n",
       "[29,]   0.000000      0.000000      0.000000   \n",
       "[30,]   0.000000      0.000000      0.000000   \n",
       "[31,] ...           ...           ...          \n",
       "[32,]    0.0000        0.0000        0.000000  \n",
       "[33,]    0.0000        0.0000        0.000000  \n",
       "[34,]    0.0000        0.0000        0.000000  \n",
       "[35,]    0.0000        0.0000        0.000000  \n",
       "[36,]    0.0000        0.0000        0.000000  \n",
       "[37,]    0.0000        0.0000        0.000000  \n",
       "[38,]    0.0000        0.0000        0.000000  \n",
       "[39,]    0.0000        0.0000        0.000000  \n",
       "[40,]    0.0000        0.0000        0.000000  \n",
       "[41,]    0.0000        0.0000        0.000000  \n",
       "[42,]    0.0000        0.0000        0.000000  \n",
       "[43,]    0.0000        0.0000        0.000000  \n",
       "[44,]    0.0000        0.0000        0.000000  \n",
       "[45,]    0.0000        0.0000        0.000000  \n",
       "[46,]    0.0000        0.0000        0.000000  \n",
       "[47,]    0.0000        0.0000        0.000000  \n",
       "[48,]    0.0000        0.0000        0.000000  \n",
       "[49,]    0.0000        0.0000        0.000000  \n",
       "[50,]    0.0000        0.0000        0.000000  \n",
       "[51,]    0.0000        0.0000        0.000000  \n",
       "[52,] 2544.1561     2055.1371     1661.081352  \n",
       "[53,]  716.2475      699.6478      811.088694  \n",
       "[54,]    0.0000        0.0000        0.000000  \n",
       "[55,]    0.0000        0.0000        0.000000  \n",
       "[56,]    0.0000        0.0000        0.000000  \n",
       "[57,]    0.0000        0.0000        0.000000  \n",
       "[58,]    0.0000        0.0000        1.178908  \n",
       "[59,]    0.0000        0.0000        0.000000  \n",
       "[60,]    0.0000        0.0000        0.000000  \n",
       "[61,]  270.1759      287.8694      278.222285  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Determine the size factors to use for normalization\n",
    "dds_smoc2 <- estimateSizeFactors(dds_smoc2)\n",
    "\n",
    "# Show Size Factors\n",
    "sizeFactors(dds_smoc2)\n",
    "\n",
    "# Extract the normalized counts\n",
    "smoc2_normalized_counts <- counts(dds_smoc2, normalized = TRUE)\n",
    "smoc2_normalized_counts "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hierarchical heatmap by condition\n",
    "When performing quality assessment of our count data, we need to transform the normalized counts for better visualization of the variance for unsupervised clustering analyses. To assess the similarity of the `smoc2` samples using hierarchical heatmaps, transform the normalized counts and perform hierarchical clustering analysis. Assume all libraries have been loaded, the DESeq2 object created, and the size factors have been stored in the DESeq2 object, `dds_smoc2`.\n",
    "\n",
    "- Transform the normalized counts from the dds_smoc2 object using the vst() function with the blind argument and save to vsd_smoc2.\n",
    "- Extract the matrix of transformed normalized counts from the vsd_smoc2 object using the assay() function and save as vsd_mat_smoc2.\n",
    "- Calculate the correlation values between samples and save to vsd_cor_smoc2.\n",
    "- Create a heatmap of the correlation values using pheatmap() with an annotation bar designating condition from the smoc2_metadata data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAABYlBMVEUAAAAA2uBFdbRJebZNTU1OfbhSgrtXhr1ci79gj8JllMRoaGhpmMZunclzoct3ps58fHx8qtCAr9KFs9WKuNeMjIyOvNmTwNuYw92ZmZmampqcxt+hyuGmzeKnp6er0OSw0+aysrK01ui52em9vb2+3OvD4O3I4+/M5vDQ0NDR6fLW7PTXMCfZNSrZ2dnbOy3b7/bdQDDfRjPg8/fh4eHh8/TiTDbj9PHkUTnl9e3mVzzn9eroXT/p6enp9ubrYkLr9+PtaEXt+N/vbUjv+Nzw8PDw+djxc0vy+tX0eU70+9L2flH2+874hFT4/Mv6ilf6/cf8j1r8lF38mWH8nmT8o2f8qGv8rW78snH8/cT9t3X9vHj9wXv9xn/9y4L90IX91Yn92oz+34/+4ZL+45X+5Zj+55v+6Z7+66H+7aP+76b+8Kn+8qz+9K/+9rL++LX++rf+/Lr+/r3+/sD/kon////OqwQtAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dh3/8Wn7XYREusLBZCBtC4ktYEgg9LG3pLBDCzbIECL333vv8/1zbU6Qjjccz5zOWv6Pn/Xrdn+2xJR+NznNV3IadpO6GtQcgPUIgSYFAkgKVgjRsrrWfcb23Uruq1GATbW6D61ZqV5UabKLNbXDdSu2qUoNNtLkNrlupXVVqsIk2t8F1K7WrSg020eY2uG6ldlWpwSba3AbXrdSuKjXYRJvb4LqV2lWlBptocxtct1K7qtRgE21ug+tWaleVGmyizW1w3UrtqlKDTbS5Da5bqV1VarCJNrfBdSu1q0oNNtHmNrhupXZVqcEm2twG163Urio12ESb2+C6ldpVpQabaHMbXLdSu6rUYBNtboPrVmpXlRpsos1tcN1K7apSg020uQ2uW6ldVWqwiTa3wXUrtatKDTbR5ja4bqV2VanBJtrcBtet1K4qNdhEm9vgupXaVaUGm2hzG1y3Uruq1GATbW6D61ZqV5UabKLNbXDdSu2qUoNNtLkNrlupXVVqsIk2t8F1K7WrSg020eY2uG6ldlWpwSba3AbXrdSuKjXYRJvb4LqV2lWlBptocxtct1K7qtRgE21ug+tWaleVGmyizW1w3UrtqlKDTbS5Da5bqV1VarCJNrfBdSu1q0oNNtHmNrhupXZVqcEm2twG163Urio12ESb2+C6ldpVpQabaHMbXLdSu6rUYBNtboPrVmpXlRpsos1tcN1K7apSg020uQ2uW6ldVWqwiTa3wXUrtatKDTbR5ja4bqV2VanBJtrcBtet1K4qNdhEm9vgupXaVaUGm2hzG1y3Uruq1GATbW6D61ZqV5UabKLNbXDdSu2qUoNNtLkNrlupXVVqsIk2t8F1K7WrSg020eY2uG6ldlWpwSba3AbXrdSuKjXYRJvb4LqV2lWlBptocxtct1K7qtRgE21ug+tWaleVGmyizW1w3UrtqlKDTbS5Da5bqV1VarCJNrfBdSu1q0oNNtHmNrhupXZVqcEm2twG163Urio12ESb2+C6ldpVpQabaHMbXLdSu6rUYBNtboPrVmpXlRpsos1tcN1K7apSg020uQ2uW6ldVWqwiTa3wXUrtatKDTbR5ja4bqV2VanBJtrcBtet1K4qNdhEm9vgupXaVaUGm2hzG1y3Uruq1GATbW6D61ZqV5UabKLNbXDdSu2qUoNNtLkNrtuau2q4eytu3Adt4Wd/BjbTqpDKfwID0L7SkD79/69Xn8erD2Az1Ya09gBW/wSffgCbCaR7dv9P8NmPyZsJpK4VrD2PV38GtA8kA1AgkAygeJ9jG0AygNp9kus8kAygdMMn2QaQDKByw2fZBpAMoEA/3zZ63+fYhs8xCunNtg3pSeW74/S4pvtDer5n8fzf937i8NaVy0dGsdzT/+usdwX9A/hVfa29fP8A7jg9rukPtY3el4X0Sgik6PIgbQvS+OXVy0dGsRxIay//MJD+cNvofe+Zwt/9seGLn31+5ftfDsOX339eanh+9Tsv7/zm8M3v7Y9G+2/ceuG09LFnA+nNFaw9jzuX/zhIvZcYF+qE9L0XHz+32331xfMrX3z1PMSXV788vHMOaeljzwfSmytYex53Lv/xkG69xLhQJ6Qvvz6cfG/45m73nWcNz299PcQvv9r93PNAX978znwDlj72fCC9uYK153Hn8h8Lafwy3B9tu27x46i+GL4+wHw1fPH80Fe7A52Xx2aQlj72jU9xw1a9N5DWXj4F6f6XGBdKQToxOffa2x/xxqe4bkRXBdLay4cgfcAlxoX+WNt1iy8ekQ7v+OIdR6TJSpY/xdXb9P5AWnv5EKQPuMS4UCek52F8t92Aw0DfswGHx84H0psrWHsedy4fgvQBlxgX6oR0PKR+f3SYPAx0ckj9etjPfp5fW/rY84H05grWnsedy6ch3e8S40J/su3K5b++jptd5B0H9fU7f+y7h4H+/i+Gn5hf5F3eAJDeXMHa87hz+XsekQ7vyFxiXKgX0v0D6c0VrD2PO5cPXiPd+RLjQn+6rXduxgPpzRWsPY87l0/ftbvfJcaFQFoVAkipryPd/RLjQhuHpPLdcXpc059tW3tAs+4JSQoFkhToz7ddtfRHHHFBUoE6IV26BgwEkgq0bUgfdkWsu3XH6XFNf7HtqqXLQ5r97pcre/rlvnqX/+WnH+nr6Rt9PX27s6dv9QXSewMJpAKQfqHtqqVBAgmkl0ACCaRAIIEEUqBfbLtq6cuQhtc/anFJwxvvBwkkkI4/b3X7CEECqQCkH7RdtTRIIIH00n0hvfwUyOup3f4PZB9eOfy97P0fzh5GrzeBBNLmIR2vj4bJK6PH92+MX28CCaQCkH7YdtXS74d0eOvkaPTm+GNmgQQSSCCBtA1Isz1z1dLdkFwj9a4ApM/RbM9ctXQ/pMnDuyU2IL25ApA+R7M9c9XSIPU66FwepM8C6W+0XbX0DXftmsf3b7hrBxJI7zoiHb+OtH+HryOB9EiQfqXtqqV90ypIIL0EEkggBQIJJJAC/e22q5YGCSSQXgIJJJAC/Z22q5YGCSSQXuqE9AG/bgwkkB4f0vk/NxIbIUggFYD0d9uuWhokkEB6CSSQQAr099quWhokkEB6CSSQQAoEEkggBfr7bVctDRJIIL30mSAtkwEJpAKQ/kHbVUuDBBJIL4EEEkiB7gtpOP3U6+Q3re5/YHbpx2RngQRSAUj/sO34nunPfR/fmjx8EdLxx8unv2l1mP8ehwEkkB4R0vQQMfnlJKeHLx+R9v8u/RahYfl9bSCBVADSP2rbPz77hVm72W/T2oUgvbwYQALpAyAt/4Kddy89eTHpM0A6KAIJpDtDOn8p/u7FD5cnTZ8AklM7kD4IUjN9r+5wv2zhXf+kbfFTggTS54Z0vPHb3g7eTV9pZ9QN95V7IJ1uuJ3uvL1MxPdCmvy3BMk1EkgdkI5zaTpRx1PtNI8mkCbvP83u8/eVz0zUf9o2+WSjT77/03unV14m4rshTX/T6uwaafjAI9JwCqRHgtROrDPnUZMJdf05UyekyVvvhpQoD+n0KkgbgrQ4od4F6eXF6I7Y7ZDO2AYJpM8BqbmI2b0BaUnVJUgnRcN8Fa/9s7bxZ5ieVi6cbYIE0ueAtN+xlyE10+naU7vDRcFsVp2FdLphsX9j2E1eeZ2IIIFUBdLiweT6a6SFlTz3z9uumpMggfQpIE2m/OTUaXoeNZtMJzaTBRYgDSCdXyFIDwJp8RppN/s60vy8bHrW1T5y+Ke5r3wHSBV/0ypIDwhp9f5F29oDmgUSSCAFAgmkNKTh7M23m/uXbblVhwIJpAJHJJBAAikQSCCBFOhfta09oFkggQRSIJBAKgDpX7etPaBZIIEEUiCQQAIpEEggFYD0b9rWHtAskEACKRBIIBWA9G/b1h7QLJBAAikQSCCBFAgkkApA+ndtaw9oFkgggRQIJJAKQPr3bWsPaBZIIIEUCCSQQAoEEkgFIP2HtrUHNAskkEAKBBJIBSD9x7a1BzQLJJBACgQSSCAFAgmkApD+U9vaA5oFEkggBQIJpAKQ/nPb2gOaBRJIIAUCCSSQAoEEUgFI/6Vt7QHNAgkkkAKBBFIBSP+1be0BzQIJJJACgQQSSIHuCknlS0+PG/tvbWsPaNZdIXUfEHqPaL/S2dOP9tV7RHn66c6efrIvkN4bSCAVgPTf29Ye0CyQQAIpEEgggRQIJJAKQPofbWsPaBZIIIEUCCSQCkD6n21rD2gWSCCBFAgkkEAKBBJIBSD9r7a1BzQLJJBACgQSSAUg/e+243uGYTyFj29NH/6AQAKpMqRhN55yx7emD39EIIFUGNIw+nf01vThDwkkkApA+j9t+8dBAgmkK7oSklM7kEBa6v+27R8/B8nNBpBAWuh9kF4PRI5IIIF0pndCejkQuUYCCaQz/eO2/eNLYlaENCQ7rRskkEK9D9Lqd+2SnxEkkPL9rbbDOz7XF2RBAqkopOPtueOp1fThDwskkEpD+iyBBFIBSH+zbe0BzQIJJJACgQRSAUh/vW3tAc0CCSSQAoEEEkiBQAKpAKS/1rb2gGaBBBJIgUACqQCkv9q29oBmgQQSSIFAAgmkQCCBVADSL7WtPaBZIIEEUiCQQMpBGrq+6XqYvJj0V9pu/yx3CiSQYpA6fwpo//MQIJ1W+BpI24LU+3Op+7kI0mmFr4H0KJD2p2zD68/LHX+WbvGn6MaTaTid7O1fDvvfT9KuaPSrf5Zm419u65+o4UAC6SKkwyQZZr/uauHnuieQJu8fTo/MVjQSBNJ+hcdfgwLSI0GaTJbxLxgZ5h8zmV3nPvjMihZn419q65mid+kekA6vgrQhSJMJNX3jAqSXFwNIsxWC9GiQxhcxFyAtqboE6aAIpOkKQXo4SPtJchlSM5Nip3Z/oe2GmXnfQAIpAmlxIoHUs0KQHg3SZMqfzsjaN+bz6MRmssACpAvXSH+u7fYJeqdAAunyEWnpGmk3+zrS7HdWT9gMQ/vI4Z9h8qvvQTqsEKSHg7R6IIEEUqA/07b2gGaBBFIa0sIfJukNJJC2B+kO/am2tQc0CySQQAoEEkggBQIJpAKQ/kTb2gOaBRJIIAUCCaQCkP5429oDmgUSSCAFAgkkkAKBBFIBSH+kbe0BzQIJJJACgQRSAUh/sG3tAc0CCSSQAoEEEkiBQAKpAKQ/0Lb2gGaBBBJIgUACqQCk39e29oBmgQQSSIFAAgmkQCCBVADS721be0CzQAIJpEAggVQA0u9pW3tAs0ACCaRAIIFUGlLzq13P/c7XuwcSSAUg/e62wztOv0989taHOgIJpMqQRr9K/MxbHxVIIBWA9Lva9o+/AeljHYEEEkiJQALpISF9sCOQQKoA6Xe27R9v0LzcbAAJJJDO9E5Iu/2fLNuBBBJIC/1M2/7xJTQggQTSmd4HafTWRzu6A6TTF5SfVL7gzOjpHKSzX5CtD2mXPCL1HlA6j2g///Stvp5+vK+n2QS6sqef6uuzQPodbcf3DOObC6fvCwIJJJDmnYf0WQIJpAKQfnvb2gOaBRJIIAUCCSSQAoEEUgFIs/v6aw9oFkgggRQIJJAKQPptbWsPaBZIIIEUCCSQQAoEEkgFIM0GtvaAZoEEEkiBQAKpAKTZt6WvPaBZIIEEUiCQQAIpEEggFYD0W9vWHtAskEACKRBIIBWA9Fva1h7QLJBAAikQSCCBFAgkkApAmj0zaw9oFkgggRQIJJAKQJr9xr+1BzQLJJBACgQSSCAFAgmkApB+c9vaA5oFEkggBQIJpAKQflPb2gOaBRJIIAUCCSSQAoEEUgFIv7Ft7QHNAgkkkAKBBFIBSLNndu0BzQIJJJACgQQSSIFAAqkApNmfuV57QLNAAikHaRj9Ke4bJ87yGkACaUOQpn9i/KaJc2YVv6Ht9s9yp0ACKQVp9OfFb50456YjSCA9AqT9KdvweuJ1OPs6noVNTsfGk2k4neztX76uYr6i8ZEIpNEqnwPpQSAdJsnrdB+fgR2nzzDf/YePnXzw/pHZio6Ho+XZ+OvbOudovntAOgTSI0GaTJbjtB+/Mnk5euPMBy+vaPlmwyYhHQNpQ5AmE2r6xgVILy+GYbds8bVvtN06J+8WSCBdhDS5iHkb0pKqS5AOiobllTwHEkiPAGk/SS5DambSDad2C2vZgQTSZiAtTqSrIU2vt079urbOqZkPJJAuQppM+dMZWfvGfB6dXEwWWIB0OLsbllYDEkgPAWnxGmk3+zrS/mZtO7smH3zmGmm4cNfu17ZlZmgwkEC6DGn1QAIJpEAggbQ9SMevyOem0q9py606FEggFTginYfUXJQtfvvfRwQSSAUg/eq2wzumd/mmNxI/lBJIIBWGNP2609JXqz4qkEB6VEgfGkggFYD0I237x89Cco0EEkjzroV05hsk7hhIIBWAdK7mYuhwj8E1EkggXVMrZvzNRiCBBNI7WxIDEkggXZnb3yCBlOitL8h++EBAAqkqpNOv9Nq/4VuEQAKpcCCBBFIgkEACKRBIIIEUCCSQQAoEEkggBQIJJJACgQQSSIFAAgmkQCCBBFIgkEACKRBIIIEUCCSQQAoEEkggBQIJJJACgQQSSIFAAgmkQCCBBFIgkEACKRBIIIEU6K6QVL77zIwH7K6QZr+x+cqefrSv3gPKt3oPaU+/0NfTDzrrXQFI7w0kkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIpB2n4uvaRl9l1aYa177/TjLxjIIEUgzQnMzQv37sikMaBtC1Iw+jfyUPXzrDBEWkSSI8CaX/KNuxeXh7O347ncZMTuvFkellutOTpldE6jsex4fgxIB3W+hJIDwLpOL1fCexOJ3HH6XOaQrMj0nShYfr4Yc3NVARptFZHpIeCNJksx4PG+JXJy9NbJ2fNBw/TpYZmuVqBBFIE0nTXj98CqXOtID0KpOmVzP6fZUjt3e/d25AWr5HmqykQSCBdhrTfp5chtTPpIqR2Ha6R5msFaUuQlicSSIG1gvQokCZ2jv8M7RsL8+j4cVNRs+XbVYA0WitIjwJp8RppN/s60usXPWaXTK//nb5GNF7ENdI71grSw0DSxUACCaRAIIGUhjQMCyd5jx5IIKUhbTKQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgeKQRn8Z50nlC86Mxy4P6bTCp2909vTtvp5+vLPuI0rvEe0XO3v6YV8gvTeQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgmkHKTjH7QfPfIyGS7NsOn756v5/IEEUgzSnMzQvLxxNQUCCaQUpGH07+Sh62bYwmoKBBJIlyHtz7WG3cvLw4nX8QRsciY2nkwvy42WPL0yWsfxADQ5owPp5Z+XQHoQSIdJsiewO519HafPGQHHj5u8Ml9+sq75agp0H0ivgfRIkCaTZRidgQ3zj5nMhpOz5oOH6VKL54RlAgmkCKSF/b8DqXeFr4H0IJCmVzL7f5YhtXe/d29DOneNVM4RSCC9A9J+x16G1M6ki5DadUzfUyiQQIpAWp5It0Eq6AgkkC5Dmtg5/jO0byzMo+PHTUXNlp+vq1oggXT5iLR0jbSbfR3p9Yses0um1/+OX0eaLLJ0jbSwmgKBBNJlSLoYSCCBFAgkkNKQhqHm2VlXIIHkiBQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUiCQQAIpEEgggRQIJJBACgQSSCAFAgkkkAKBBBJIgUACCaRAIIEEUqDx3yzMdFr3k8q31rwsV/yPQY0hdR5Qvv300309/UxnTz/oq/eI0ntE6z6kgfTeQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkHKQpn/V/vWRlxlxaZpN3z9bS4FAAikGaU5maF7euJoCgQRSCtLQToDbIA3XLvApAgmky5D2p2zD60nX4czreAY2ORUbz6iX5UZLnl4ZreN4ABqfGIIE0uNBOhwj9gR2p7Ov49FjWJoAo4+bvDJffrKudiUluguk4TWQHgnSbjzVh9GJ3DD/mOls2J354GG61Miimw2TpwSkDUFanlC3QZqtpkAggXQR0vRKZv/PMqT27vfubUhnr5HKSQIJpMuQ9jv2MqR2Ol2E1K6j7N0GkECKQBrfKpjPhmsgTd9ZJZBAughpYmcY7+XJGwuT6fhxU1Gz5Sevzy6YKgQSSBchLV4j7WZfR9rfrF2aDaOvI00WWb5GctduB9JDQtLFQAIJpEAggZSGNAwLJ3mPHkggpSFtMpBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEB5SKe/MPWk8qWnx8N2l7+qdoD0rc6efrKvp5/qrPuI9MO+eg8o3Yc0kN4bSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCaQcpNMf4j49Mv4795dmzNm3P38ggRSDNCczNC/fuyKQxisFaVuQhvHOn8yEK6fZ4Ig0WSlIjwJpf8o27F5eHs7fjudxkxO68Yx6WW605OmV0TqOx7HjiSFI05WC9CCQ9gR2ewK700nc4T2jeTQ7Ik0XGqaPH9bcrAuk15W+BtIjQdqNpvppro9fWZhQEzDtBw/TpYbRQiCNA2lDkJYn1C2Q3nOX7/MFEkgXIU2vZPb/LENq737v3oa0cI00LKymQCCBdBnSfqZchtROp4uQ2nUMh+uCapRAAikCaflAcgOkybvqBBJIFyFN7Exu1E3eWJhMk5t1o4ufdvl2FSCNA+lBIC1eI+1mX0daOCc7gTl+HWmyyPLXkUCaBtKjQNLFQAIJpEAggZSGNBS98dYVSCA5IgUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKBBJIIAUCCSSQAoEEEkiBQAIJpEAggQRSIJBAAikQSCCBFAgkkEAKdFdIKt8dp8djtam/qibdK5CkQCBJgUCSAoEkBVoH0vNn7fnz8b3LG0D/CjRplWdyGP23xvIG0L8CTQPJABQIJANQIJAMQIFWutkw9O3E3uUNoH8FmuSZlAKBJAUCSQq0zs2GQystbwD9K9C0NZ/I3s/dPXYD4CjVqs9k+XlkANoHkgEoEEgGoECr3mxYa3kD6F+BprnZYAAK5NTOABQIJANQIJAMQIHcbDAABfJMSoFAkgKBJAVaCVLv+Xn3+b0BuESKttKPmnd+7t7lDaB/BZq0zl27zk/eu7wB9K9A00AyAAUCyQAUyDWSASiQu3YGoECeSikQSFIgp3YGoEBuNhiAArn9bQAKBJIBKBBIBqBArpEMQIHWumvX+al7lzeA/hVonKdSCgSSFGjVmw23fvbe5Q2gfwWatg6kzt3Yu7wB9K9A01Y6IvXtxt7lDaB/BZq21qld19cxepc3gP4VaNpq10hDx+fvXd4A+legaevdbHj95+Zp1LO8AfSvQNNWvGt3+7fx9y5vAP0r0DTPoxTo4yH1fsbuERuA/3vmA8kAFAgkA1AgkAxAgUAyAAUCyQAUCCQDUCCQDECBQDIABfKcSoHW+V67Qyst37+Cw3o6l3dwepjWfCLrzqJh6KTYu4LVB6CmVZ/GupJOy/etoncFqw9A+0DqXB4kPQeSFAikm5Y7XV90XuPcuILVB6CmVe/arbV8/wqkaeaSFAgkKdBKkIr/BdXRV2G6vw7Ud4201gDUtM6zOHR+7t7lDaB/BZq0zs2Gzk/eu7wB9K9A00AyAAUCyQAUyDWSASiQu3YGoECeSikQSFKgVb/XLvA9owaw0go0bc0nsuw3fxuA2vwYhQEoEEgGoEAgGYACeSKlQO7adazh5qUDK1h9AJrkrl3Hkr5FSIdcI/Us55tWtQ+knuVA0j6QepYDSftWvdmw1vKBFYz+XWUFqw9A09xsuHHRtW+arT4ATXJqJwUCSQoE0s0Ldp0Z9a5g9QFompsNNy0/+m+VFaw+ADV5Im9p9Xm8+gDU5Im8pdXn8eoDUJMn8pZWn8erD0BN6zyRQ+cfuOpdPjOANVew+gA0bZ2bDa//9f7vtPOel2mkXGtCcmKjhwkkKdBqkHpOrXqX7x/A6QtRt62idwWrD0BNK91s2LlI0kPlOxukQOaSFAikm3o5NfRNqzq26qndrWdXvcv3D2D0320D6FzB6gNQkyfyllafx6sPQE2eyFtafU9tRu4AAADsSURBVB6vPgA1rfVE1j47X30erz4ANa0Gad1deHDc8RXZjqUDK1h9AJq2HqTAQanvplfnKqRRK0IKUOr6HqPjNwpJ3a0Kqf/T37g8SLvql6mfrY1+HWm/mt1akLpPLXsv8narX6Y+WCsfkdbq8E2rnf9X7vhuwc5VBC7yIpep2rdRSKnvjrn9Iq1PQuLcNHOZqtc8j33dfkAZTq/e+mkDkMyAUJ7GW4pcpPVM5O6LvO4t0DTP4iqdjiV9Enu/gVypPJurdDiidP1wY2QMyuTZlAKBJAUCSQoEkhQIJCkQSFIgkKRAIEmBQJIC/X+0zaAWSL5xhwAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Transform the normalized counts \n",
    "vsd_smoc2 <- vst(dds_smoc2, blind = TRUE)\n",
    "\n",
    "# Extract the matrix of transformed counts\n",
    "vsd_mat_smoc2 <- assay(vsd_smoc2)\n",
    "\n",
    "# Compute the correlation values between samples\n",
    "vsd_cor_smoc2 <- cor(vsd_mat_smoc2) \n",
    "\n",
    "# Plot the heatmap\n",
    "pheatmap(vsd_cor_smoc2, annotation = select(smoc2_metadata, condition))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA analysis\n",
    "\n",
    "To continue with the quality assessment of our samples, in the first part of this exercise, we will perform PCA to look how our samples cluster and whether our condition of interest corresponds with the principal components explaining the most variation in the data. In the second part, we will answer questions about the PCA plot.\n",
    "\n",
    "To assess the similarity of the `smoc2` samples using PCA, we need to transform the normalized counts then perform the PCA analysis. Assume all libraries have been loaded, the DESeq2 object created, and the size factors have been stored in the DESeq2 object, `dds_smoc2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAP1BMVEUAAAAAv8QzMzNNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enr6+vw8PDy8vL4dm3///92l2KZAAAACXBIWXMAABJ0AAASdAHeZh94AAAXF0lEQVR4nO3diXbiSJRFUWUDBo80Xfz/tzZiMrMRXKwIs/daVeXxlRzJSUDIUjMH7tb0vQHwFwgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUHAUUhvL00zH331sS1QrYOQZsNmYd40n/1sDtTpIKRxM1lUNH9vRv1sDtTpIKRFRNt/gGsJCQJOP7SbNON+NgfqdLizYdAsDab9bA7U6egh3OuwaYaTWR/bAtW667nQ/56x+MR/C+c+faOz/7sCp1a1sVUvQSqEez0mpP/WHrqEJU+tamOrXoJUCPc6DOll+YFmeNdzpP+27pkC9TgIabLa791ct9fu3N8a3yE98u+ikqdWtbFVL0E+idschDRYHRv0dd3rSGd+1v/+e0hJbkVVTX3ukDYBCanIsTVNfe6QXprxbD6fTa471u7Mzyqkuja26iV4RBS3OAhpunlB9qrfozjzswqpro2tegkeEcUtDh/CzSbLF2Sv22l37oe1s6GqjS17Cf4tXJiaT+I2D3kdSUhVbWzJS/Bv7ezUVAj38oLsQ6ZWtbEFL8G/f4cl1RLSZP0k6Z69dg4RqmtjC16CakOaNE0mpAeoaWpVG1vuEvz7d1RSJSENmrcO3/zAJax8alUbW+4S1BtSt9+MfeASVj61qo0tdwnqDemlufybSIPW9r0HLmHlU6va2HKXoN6QpoPRxZeQBnvvPXAJK59a1cYWvATV7mxoftjZIKRex9Y0VUgXQtrvSEi/PbamqV6QveD7KdL/tB6xQbCnzajvbbhCx5C2/1p66N9FVU+tamOrXoJsDrc7OrLh5xdkhdTb2JqmPndI1xzZIKTextY09blDGjRfo2Y6G525GoWHdv2OrWnqc4e0uCd6bT7mszO/ITuY7+2563cJS55a1cYWswQHv3p01dTHpdHNcUgf7eF2Z19H2j2wQUi/PbamqZ3HHu3pvmrq49Lo5ugQofdpM5x/Ovq7yLE1Te069vi116umPiKKWxwE0xY0avc13HVeu5r+vMu4Ff3BqUWENB01w4/VdYq+BounK9Px4qbdHgS3OX/j8p+X5vKBcdc4vOf5GLbXdmkmV31zaAmvVNPUqja2iCU4cXzqVVMv3kA3F1dpcxkt7hxW7w9m+yGN1x+7y6NOov8ANU2tamOLWIKHhPTajOaz0SqX9r5heZq5Ufvmbkij2epjdxHSQ6ZWtbFFLMFDQho2i4ds01Uu0+/3h/shfa0/dpfdkJaXM/er5gWPrWlqCSHtPRM68/7uZ+4gpIdMrWpjy1iCR+xs6CmkzkJLeKWapla1sWUswSNC2n1ot/P+5qHd94O+6XXn6L7gIKRRp4sw37CEd5ym6y/fiv7g1BJekF3vXNiG9L2zYdC8b3dDLHdIvHYM59DRsXZdvrnzEt514sg/fSv6e1NLOERod/f39/vtru7l0dmvmx3ji4917ObIQThfoytP+73UdQnvO5XxH78V/bWpRRy02r4g+/4d0vcLsu2ZUF/Xz5EWXzROvyDbbWdDVy6JSQ8u392Ebup3hdTx76I7L/fir+OqppZwj9S0vw40uXy822NC6qbjEgrpmaaWENL691QvPm4TUkbVt6KSp5YQ0vxt2Pz0/OdBIV1xzoZvHZdQSM80tYiQfs+vXo3CXrsnmvrcIf1wzoYDXZdQSE809blD+uGcDQc6L6EXZJ9n6tOHdPGcDftuWEKHCD3L1OcOyTkbih5b09TnDsk5G4oeW9PU5w7JORuKHlvT1AJC+r8Lbg3mnMODVjt9c79LWPLUqja26iW4dPvsMaRm+HHxy50gstexNU197pCGTTN4PX9mIuf+7ndsTVOfO6T5dDJompdzL8cKqd+xNU198pAWPidNM3w/+dVC6ndsTVOFtDwpxIVryC7/5dKXFK/ve6Tx4h7p7eRXu0fqd2xNU5/8Hmn5HGnsOVKZY2uaWlhI7bFpvxjSsL0zsteu1LE1TS0qpM3R0r8WUvNy8XUkIfU7tqapJYX0/fs7J0NqTzH8wy+L/3js6cEX/HBxCyH1O7amqfWE1FzZykUdv9uRDb2OrWlqQSHtnuOgkJD29buEJU+tamOrXoJLt8/rQlqeWGH50G71qk+zPm/k5uz66w9+f0hIvze1qo2tegnuD2m+DqU5+O98+8Sp2XtTSL84taqNrXoJgiEdtDPf3QPx084IIVV9Kyp5akEhXbGz4VxIm5Mu7Lx52onPvQ6awekDG4TU89iapv6RkHYf6l1K6fgzr8vT2l11uZh+l7DkqVVtbNVLcFVIl1+Q/SGkw4d414Y0aA/8/rjqejH9LmHJU6va2KqX4LqQLh4idLizYTekm3Y2rH47tmn//e4sQiWOrWlqYSEdOhHSvNl7FHf77u/VryFNPLQrdmxNU+sJKWI3pNmiocHbrN3ZcN0VNftdwpKnVrWxVS/BpdtnXyG1KS0imvxwwB3Uob+QFt4GTXNtSv3+XVTy1Ko2tuoluHT77C2k1StI78OmebnqBHdX/bB3nO378hJmVH0rKnnqE4e0fQXpYxg79/dd15+4vIQZVd+KSp76xCHtvIL0+XLNN//8w953RaTLS5hR9a2o5KlPHFKHV5BWfv5hhfS0U584pA6vIK38+MPuHnYbX8KMqm9FJU994pDm17+CtPLjDyuk5536zCF19eMPK6TnnSqk6/34wwrpeac+c0ifw80pVlO7v+1seNqpTxzS53Jnw/J65kIqcmxNU584pFF7xcuPQVuSF2SLHFvT1CcOaVXPtC0peFXznzK6OjO3oqqmPn1Ii5JezoY02DtDZGAJO9xjuRVVNbWwkP4t/FZI4/XFzL+a0dmQ9t67fwm7PIdyK6pqalEh/Vv7nZC21xd7P/s7tULqdWxNU0sK6d+/o5LuruXSp6bjdSifo9PfcnBKlLuXsNPrTG5FVU195pB+9P0UKXPpy92Q7h4G+447+i7ppoG5kLb/Wrr77yL3SH93akH3SD+EtDl90PcZg5Zn1W/W59Lf+0wopO+ahNTH2JqmVhTS4Rkit+cnPvrM/SHt7PZO7v62s+HPTq0opN1Mjk+1et159Pt9aCekvzu1oJB+2NlwTUirx3bRkLJX7POC7F+d+qdC2n2IlwjpEZe+dIjQ35xaUkiXX5D9OaTuz5E2L8QGj7V72BKWPLWqja16Ca4L6eIhQteF1O2hnZCKHlvT1MJCOnQipL2d3M3hZ7IP7YTU69iaptYTUoSQHjK1qo2tegku3T6FVP3Uqja26iW4dPvsLaT8ORseuYQlT61qY6tegku3z75CesA5Gx65hCVPrWpjq16CS7fPvkJ6xDkbHriEJU+tamOrXoJLt8++QnrMORsetoQlT61qY6tegku3z35DunjOhn39LmHJU6va2KqXIB3ErTqeswE4peM5G/b1+3dRyVOr2tiql+ARUdyi2zkbDvS7hCVPrWpjq16CfBK38YLsQ6ZWtbFVL0EqhHsJ6SFTq9rYqpcgFcK99kN6W11AdnzVNc2F9Otja5r6zCGNmmaZ0GC9++4n/S5hyVOr2tiql+ABTdxkN6T3ZvCxfONz0Lxf8839LmHJU6va2KqX4BFR3GL/EKGP9VsfqyPuftLvEpY8taqN7WsJ2l9avXtqPonbnDiy4eDNC25ewpvUNLWqje1nCTanUbhzajqIW50LaXD8pcduXMIb1TS1qo3tZQm+T+xz39R0ELfaf2g3Xb81bcbXfPNtS3irmqZWtbFCCtgN6W2bz3j7bOmi25bwVjVNrWpj+1iC3ZOf3jU1n8Rt9p4LDZqXz8V/Pl+a4dEXrh7rPeC8djcvYclTq9pYIQXsH2s3aFYG08OvG3yfrTh66cs7lrDkqVVtrJACDvbOvb8sMno5fhFpMBdSAWNrmvrcIZ0npALG1jT1iXc2XHQQUuaKfTyt75D63pKMvZC+Rk0zPnp6tOIeqYCxNU194hdkv1Z7Gk4f+n0qpN9V1R1gTRvb17bedHdU6sIenbNhvP9a7HZ/t5A6qWlja9rWYjf26BCh2Zmjg4TUSU0bW9O2Fruxx8fanTleVUid1LSxNW1rsRvbLaSDIxt+V6lLeFJNG1vTtha7sVeHBJwnJAjYD2lHb1sEFRISBAgGAoQEAZWEtNntPuhz//v16tjKypZ1s5FlbmsdIW1fCC5wBU/o9XXrDupa1sHOf8rbYCE9QKF/1kfqWlYhhQxKXL+TCv2zPq2WZR3s/re8La4spDIfHx8q9M/6tFqWdfsUaT4vcXHrCangRTxUyWYuVbOsm40sdGPrCmn/jWIV+md9UkXLOhfS7fZ/sfDgrVIV+md9Sk3LOhfS/Q5/TbdolWzmvKpl9dAuYWfXZ3lreKTQP+tjNS3rYOefAre2jpAGZb+qfaSizaxnWXcPw+h7W47VERIUTkgQICQIEBIECAkChAQBQoIAIUGAkDpbn2ZpvL5qx9uoaUbbaxxODhf0a9QMJrP2rfdhM3xbf9Hgdd6eZv34Ur2n/neJjebB/Cl1tj1jWVvS5rK7o9XnJoc3+6/tNXnfl2+1Jb02H+/Na/vF11w7XkhV8KfU2fqWPVnGM1heme1jsCxkPj46IeC4/cRbe72cYfM5/1he6mOw+JrFG9fdIVEHIXW2aaX973vzsnx7Xcjgc/3JbU/r00CPdt9t/7X4Z9J8X/R6E9VwcTf3+bK4C5ssv+5rMFp9y87Hpi+rt+bbN2aLfsezh/28XENIne2G9LK4m1laPmGabD95GNL63Y/2rmlzj7R36OVL015ydLro6WP1WHE5a9SMl9+7+7HB+q3Z8o2X5bwF9279ElJnqypmy0sbHj2BOfzAqH1ot3nm9LK6/3pt3j+at7fVo8G1VWLts6Zhe0f1tbrXmqwn7n5sNFs8VBy0XzueL+8AX9svm+xN49cJqbPtzobpFSF9LvdEbJ5WDTclDV4Xd0iz9d67peHqsWH75vTjdbSKZrqdePyxYbN+NDdc3em9xH9QOhBSZ6uKBsvd3z+GNP9YxDPbfvR9eSezfmuz927pbfEg8XP53mhzFYOdh4UnPrbz6NFlD/pn+TvbvclunyPNP48/uTXd7mz4/vziDmX9XGlltnikNmnvZMbN8O1juh/NqY8JqSiWv7Pdm+xmr93nYHz8ya3P9n5osHwotrvbYS+HRSzT5bD1JbGPozn42MFDO/rlD6GzvVa2ryN9nfjkvL2Vf8xnL+1nJ+3jtrdmvPn4bL53j7R8NvW5nPA5n40OQzr+2GSR4tfmjUXRozk9ElJne61Mh9s907uf3H7N2/azq93V7TEO87aal83eu+2o4WoX9qQ5fj506mPTzV7v1eBNyfRDSJ0d3Ol8jAc7x9odhjR/2x5hN21fN111NF/eR6323m29rV+gXXzZ6PNgZ8Opj32N1vOmy89mf0g6EhIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC/h+cuxPZsna1JgAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    " # Transform the normalized counts \n",
    "vsd_smoc2 <- vst(dds_smoc2, blind = TRUE)\n",
    "\n",
    "# Plot the PCA of PC1 and PC2\n",
    "plotPCA(vsd_smoc2, intgroup=\"condition\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "python3102jvsc74a57bd055db829a23e27f1a4129460ab65de434bf1fc9efa4ae0ce0afb72d96498ad818"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
