{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 01: GenomicRanges - Basic GRanges Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01. IRanges - Basic Usage\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview\n",
    "A surprising amount of objects/tasks in computational biology can be formulated in terms of integer intervals, manipulation of integer intervals and overlap of integer intervals.\n",
    "\n",
    "Objects: A transcript (a union of integer intervals), a collection of SNPs (intervals of width 1), transcription factor binding sites, a collection of aligned short reads.\n",
    "\n",
    "Tasks: Which transcription factor binding sites hit the promoter of genes (overlap between two sets of intervals), which SNPs hit a collection of exons, which short reads hit a predetermined set of exons.\n",
    "\n",
    "IRanges are collections of integer intervals. GRanges are like IRanges, but with an associated chromosome and strand, taking care of some book keeping.\n",
    "\n",
    "Here we discuss IRanges, which provides the foundation for GRanges. This package implements (amongst other things) an algebra for handling integer intervals.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This document has the following dependencies:\n",
    "# BiocManager::install(\"IRanges\")\n",
    "library(IRanges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot Iranges\n",
    "plotRanges <- function(x, xlim = x, main = deparse(substitute(x)),\n",
    "                       col = \"black\", sep = 0.5, ...) {\n",
    "    opar <- par(mar = c(5,1,1.5,1))\n",
    "    on.exit(par(opar))\n",
    "    height <- 1\n",
    "    if (is(xlim, \"Ranges\"))\n",
    "        xlim <- c(min(start(xlim)), max(end(xlim)))\n",
    "    bins <- disjointBins(IRanges(start(x), end(x) + 1))\n",
    "    plot.new()\n",
    "    plot.window(xlim, c(0, max(bins)*(height + sep)))\n",
    "    ybottom <- bins * (sep + height) - height\n",
    "    rect(start(x)-0.5, ybottom, end(x)+0.5, ybottom + height, col = col, ...)\n",
    "    title(main)\n",
    "    axis(1)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic IRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 3 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         3         3\n",
       "  [2]         3         5         3\n",
       "  [3]         5         7         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify IRanges by 2 of start, end, width (SEW).\n",
    "ir1 <- IRanges(start = c(1,3,5), end = c(3,5,7))\n",
    "ir1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir2 <- IRanges(start = c(1,3,5), width = 3)\n",
    "all.equal(ir1, ir2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An **IRanges** consist of separate intervals; each interval is called a `range`. So ir1 above contains 3 ranges.\n",
    "\n",
    "Assessor methods: `start()`, `end()`, `width()` and also replacement methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>3</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 3\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 3\n",
       "3. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 3 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "start(ir1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 3 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         1         1\n",
       "  [2]         3         3         1\n",
       "  [3]         5         5         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "width(ir2) <- 1\n",
    "ir2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 3 ranges and 0 metadata columns:\n",
       "         start       end     width\n",
       "     <integer> <integer> <integer>\n",
       "  A1         1         3         3\n",
       "  A2         3         5         3\n",
       "  A3         5         7         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# They may have names\n",
    "names(ir1) <- paste(\"A\", 1:3, sep = \"\")\n",
    "ir1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# They have a single dimension\n",
    "dim(ir1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3"
      ],
      "text/latex": [
       "3"
      ],
      "text/markdown": [
       "3"
      ],
      "text/plain": [
       "[1] 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(ir1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 1 range and 0 metadata columns:\n",
       "         start       end     width\n",
       "     <integer> <integer> <integer>\n",
       "  A1         1         3         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Because of this, subsetting works like a vector\n",
    "ir1[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 1 range and 0 metadata columns:\n",
       "         start       end     width\n",
       "     <integer> <integer> <integer>\n",
       "  A1         1         3         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir1[\"A1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 6 ranges and 0 metadata columns:\n",
       "         start       end     width\n",
       "     <integer> <integer> <integer>\n",
       "  A1         1         3         3\n",
       "  A2         3         5         3\n",
       "  A3         5         7         3\n",
       "             1         1         1\n",
       "             3         3         1\n",
       "             5         5         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Like vectors, you can concatenate two IRanges with the c() function\n",
    "c(ir1, ir2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normal IRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A normal IRanges is a minimal representation of the IRanges viewed as a set. Each integer only occur in a single range and there are as few ranges as possible. In addition, it is ordered. Many functions produce a normal IRanges. Created by `reduce()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 4 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         4         4\n",
       "  [2]         3         4         2\n",
       "  [3]         7         8         2\n",
       "  [4]         9        10         2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir <- IRanges(start = c(1,3,7,9), end = c(4,4,8,10))\n",
    "ir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAeX0lEQVR4nO3d3W6eR3JF4U+SLf+MJd//3caS7XESoDjVnZ1aLwvrORnMgYHmKm5SogTo9buk/7MX/QBpA4ckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTikJ3u9vM874aGezCG9Gx7qyRzSu+GhpACH9GR/fUf643++fHx9pl+jNzikJ/tnSB9fr0/0a/QGh/Rk/wzpD7/Qr9EbHNKT/TOkT1/pt+hNDunJ/hnSr/RT9DaH9GT/DMlvSA/nkJ7snyHRL9F/4IWezCG9G17oyRzSu+GFnswhvRte6Mkc0rvhhZ7MIb0bXujJHNK74YWkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkgPc2pNc7QDdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxkngO7YQTdagT5ipXzvZJwAumMH3WgF+oiV8r2TcQLojh10oxXoI1bK907GCaA7dtCNVqCPWCnfOxlH2sohSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBZRDeunW5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaHHFLe5P3u0IU66EaH3tt7pUdySFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQp4yJBe0lPcfQaHF3HpIc94i09MWPvEh3xgD3nGW3xiwtonPuQDe8gz3uITE9Y+8SEf2EOe8RafmLD2iQ/5wB7yjLf4xIS1T3zIB/aQZ7zFJyasfeJDPrCHPOMtPjFh7RMf8oE95Blv8YkJa5/4kA/sIc94i09MWPvEh3xgD3nGW3xiwtonPuQDe8gz3uITE9Y+8SEf2EOe8RafmLD2ie/gA5OezyFJAQ5JCnBIUoBDkgIckhTgkKQAhyQFOCQpwCFJAQ5JCnBIUoBDkgIckhTgkKQAhyQFOCQp4AlD+vnj68Pnr/Qr/pN/PSHVG3778fX68Qv9ird8/fzh2Yf++e8TXzz0AZ8dn7//WxofHhz4m68fHpDqDb8+vuKXD38+8bFj/+3vf9Hl0/eHfjz6j/nPjt9eP3799sXgR/ohb/vh8t/NmfLhw2+/f/3h9Zl+R+3H74/7/NhD//bhrxP/6/VHyz/+379O/mv+s+OHP5/w8M/TX27/Aaohv3z/LP36+kA/pPZ69qF/fn3662mfX7/+/q3oTyf/+WM+qqf2/dOXf1d+qB9fv9FP+E/++rXxU7f+x1eiv078w+vbrz5/e/1w9J//f7zpwtfXJ/oJb/n0+vLsIX18/f7Th++/SH6sn/76pd3RF/o5v/3v75lnB3/KZ8fP37+dPtVPr18e/i3z9frh++/k6Xe85edvP2348DP9jNqCIX35cPR9dNj37/JPH9K3Hzb8+NQv99/99P2HYQ9+4fsf0tcPj/6F3cdvP1V++pC+/R7py+HPbEf9/O2Xdn9s/bnfkt7/kD49+P7ffiP/7ZedTx/Sf/+fR/r4+vY7uK8P3vpf9T681yF9+fjpsX9I983r3+iX1N7BHyI8f+v/46d2X97dT+1+ffYP7N7HkH76/l3zy5NT/vmF/sl/1PXXgf9s+evZH27znxuPPv5/8+QZff/d0ddvvwH5hX5I7fPr219f+/zgv3zxvv9mw4/P/3L/3cMf+OePxB79NenT05/494k/XjyU/+x4B79u+u7pD/z10+vDc7/Yf/f9L1XTj3jD3yf+evHQh392SO+DQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkAIckBTgkKcAhSQEOSQpwSFKAQ5ICHJIU4JCkgP8CQEyTRbKJzhEAAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title \"ir\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(ir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 2 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         4         4\n",
       "  [2]         7        10         4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reduce(ir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAb+klEQVR4nO3da1tUB7ZF4UKUeAP//789UFy8xEW3nrFq7kmP90vaRMxmMJdoST85fZP0/3ZKP4D0FnhIEsBDkgAekgTwkCSAhyQBPCQJ4CFJAA9JAnhIEsBDOpDT6Q8/HF9Pp/f/frv3p9NX8Kn03/CQDuSPD+npZH55u6fz0iV5SAfyp4d0fzHvfvd27/yUdHEe0oH86SHdnE6ffvf3P51ON8gD6b/mIV3U/ancvjuv/O7m6nR1c/v4t28/XJ3efXw+pOd7ev7r3c39p5jrz9+evvH97a5+/o7ff/DT6epS75EeeUgXdb/1h6O4P537I3jw5eHvfnn839e/P6Tn73rz4ze+PL7Z+x+/48sP/vB7py+Xftf+x3lIF/V4BZ/On0zOzp85nr/x+0N6+aefv/38dh9Pp48/fseXH/z7P9GleEgX9fB55+7been3f737cN79/W9pru6P5PPVbw/p/rteff12d31+YeGnt3t5mfv7251/cF+3C/CQLurp88rDDdw9fvv9+Rvnv/n5t4f09E/v3v1z+8vb3f867vaH7/jyg9//AvDp5TxdjId0UaeXQzh9/zXay4t1vz2kn17K+09vd/ftp7+ji7H3Rf1wI99/V/RXh/TK2/36RroAe1/U88CvfrmOn/7HK4f0X73dr2+kC7D3RT0P/P3Lb2d++Mann3+J9uXxm9e//B7p879+MA/pAOx9Uc8Df3ih7sv5L9ePr8vdn8en51ftrs5/ZvTl6jev2v30dtf/frHh6V9z66t2l+YhXdTL1l/+cOjLt3/9OdKHn7/58k8//vJ2H57/2PVfh3T/yezDZd+x/3ke0kW9bP3z0z2cv1zh6Ssb3j/909unf/T0zS8/fmXDj2/3mz+QffrX+AeyF+chXdT3rZ+/gO7985/73H9yuX75WrtvXx+++emHr7W7+v5df3i7r//+EqGnH9z/R9LFeUjFrqaXFPyi1YvzkIrd/Pga3g8++3+juDgPqdjt49d6/8v755fzdDEeUrPf/17IL1kN8JCa/f5kfKkhwEOSAB6SBPCQJICHJAE8JAngIUkAD0kCeEgSwEOSAB6SBPCQJICHJAE8JAngIUkAD0kCeEgSwEOSAB6SBPCQJICHJAE8JAngIUkAD0kCeEgSwEOSAB6SBPCQJICHJAHGQzrpb13y4/d30oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUltjdsZPyldqNiY1Na4nfGT0oWKjUkv+fGT3ioPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLgIId0ko7i7xYMX8RfOshjvMZHJLzZRzzIO3aQx3iNj0h4s494kHfsII/xGh+R8GYf8SDv2EEe4zU+IuHNPuJB3rGDPMZrfETCm33Eg7xjB3mM1/iIhDf7iAd5xw7yGK/xEQlv9hEP8o4d5DFe4yMS3uwjHuQdO8hjvMZHJLzZRzzIO3aQx3iNj0h4s494kHfsII/xGh+R8GYf8SDv2EEe4zU+IuHNPmLBOyYdn4ckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkwBEO6eO709XNXfop/pMvR0j1iq8fTqcPt+mneM3dzdWxP9Afnz/Ef/GgB1jHzfm/pXF14MAP7q4OkOoVnw9f8fbq8REPe+xfn/+LLtfnB333R2+cX8fX04e7h58MPqQf5HXv//K/m3MpV1dfv929P92kn2P24fxwN4f9QH+9evoQfzndt7z/1pc/eev8Ot4/PsLBd/rpb/8DVBfy6bzSu9NV+kFmp2N/oD+erp8e7eb0+dtD0X/+5M0P814dte+j25fKB/Xh9DX9CP/J06+Nj3rr9z8TPX2I358efvX59fT+j95845n+wt3pOv0Ir7k+3R77kN6dvv1zdf5F8mH98/RLuz/6if5yvv76OfPPPuBHWcfH86fTo/rn9OngnzJPp/fn38mnn+M1Hx9ebbj6mH6M2Rs4pNurP/o8emHnz/JHP6SHFxs+HPWn+7N/zi+GHfgJ+w/p7urQv7B79/Cq8tEP6eH3SLd/+JrtRX18+KXd/a0f91NS/yFdH/jj//Ab+Ydfdh79kH78yyG9Oz38Du7uwLf+VO+q9ZBu310f9g/pHpxepJ9kVvCHCMe/9Z9etbute9Xu87FfsOs4pH/OnzVvj5zy8Sf6I/9R19MH+LHl5z/7w+38Ng79wf/Bkc/o/Luju4ffgHxKP8js5vTw5Ws3B/7ii+6vbPhw/J/uzw7+gI8viR3656Troz/i84f43V88aH4dBb9uOjv6A36+Pl0d9yf7s/MXVacf4hXPH+K7v3jQg69D6uAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLAQ5IAHpIE8JAkgIckATwkCeAhSQAPSQJ4SBLg/wCt6R1aYDVIzgAAAABJRU5ErkJggg==",
      "text/plain": [
       "Plot with title \"reduce(ir)\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(reduce(ir))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From some perspective, `disjoin()` is the opposite of reduce(). An example explains better:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 3 ranges and 0 metadata columns:\n",
       "         start       end     width\n",
       "     <integer> <integer> <integer>\n",
       "  A1         1         3         3\n",
       "  A2         3         5         3\n",
       "  A3         5         7         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAXlklEQVR4nO3d7Y4V5rGE0eHDGHNguP+7PQxOYilxS+2humpP97P+JEJCb6qags2Mkzx9B/DLntL/AYANGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAe0tPT/x7m49/8GB4Ft3lIfzOkL383LjwKbvOQ/nc0LzviWI+L27wJ3z4+MaSHxm0e0r9G8+Nfvr1/+vTz33xgSI+M2zykv4b0/seEXv7Nx7/9AgQeBbd5SH8N6Yc/vn//8OXvv5KHR8FtHtJfQ/rw/F8/hofEbR7SX0P68t8/hofEbR7SX0N6/u8fw0PiNg/pryH9z4/hIXGbh8SQ3hpu85AY0lvDbR4SQ3pruM1DYkhvDbd5SAzpreE2D4khvTXcBhBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRA4NaQnWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dboWFVfreTrfCwip9b6dbYWGVvrfTrbCwSt/b6VZYWKXv7XQrLKzS93a6FRZW6Xs73QoLq/S9nW6FhVX63k63wsIqfW+nW2Fhlb63062wsErf2+lWWFil7+10Kyys0vd2uhUWVul7O90KC6v0vZ1uhYVV+t5Ot8LCKn1vp1thYZW+t9OtsLBK39vpVlhYpe/tdCssrNL3droVFlbpezvdCgur9L2dToUFpjAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAgXJIT1By3pRDDiprpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZvq3mBlJRzr5MmXN9G8xM5KOdPJlyprp32JmJB3p5MuUNdO/xcxIOtLJlylrpn+LmZF0pJMvU9ZM/xYzI+lIJ1+mrJn+LWZG0pFOvkxZM/1bzIykI518mbJm+reYGUlHOvkyZc30bzEzko508mXKmunfYmYkHenky5Q107/FzEg60smXKWumf4uZkXSkky9T1kz/FjMj6UgnX6asmf4tZkbSkU6+TFkz/VvMjKQjnXyZsmb6t5gZSUc6+TJlzfRvMTOSjnTyZcqa6d9iZiQd6eTLlDXTv8XMSDrSyZcpa6Z/i5mRdKSTL1PWTP8WMyPpSCdfpqyZ/i1mRtKRTr5MWTP9W8yMpCOdfJmyZudNga0YEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgQYEiDAkAABhgQIMCRAgCEBAgwJEGBIgABDAgReN6QnYC3nkF71sxR4+cjTb+1lhsTLD/n0W3uZIfHyQz791l5mSLz8kE+/tZcZEi8/5NNv7WWGxMsP+fRbe5kh8fJDPv3WXmZIvPyQT7+1lxkSLz/k02/tZYbEyw/59Ft7mSHx8kM+/dZeZki8/JBPv7WX+ae/AQGGBAgwJECAIQECDAkQYEiAAEMCBBgSIMCQAAGGBAgwJECAIQECDAkQYEiAAEMCBBgSIMCQAIFXDelzaH6f3z+9+/QcePj5t6en374GHv7p/zJ1/8r/N8Mv+vrS9zf/u7/wf0jxmp6+Ztr9/ulnxneBJb37+XJoSc/vMr+Yc0P6kjr0v3f07hU/9Z//lK/vQr9NPf32/PKn4W/2lz+9vPnp6aP94Z8+puoO5f3x+9a7r9+fPz59Cj3/5en//vlP+uc3+vz0IXPZj3++Gnj83dNz5uEXf4Q+Xn1++j3x7PeXxC8Ten7NnwsKz+9e8zvIP7/Rj5ShX1L/ej/1eOaw31K/b31++px49offUh+i//Tx6TWfKf/5jb4Gfy1/f/mN6kPm4U+ZX1gfnr6lPgB8+e3pXeLj1fun77+/+/k5PuHr6z5SvupGySF9fvqSePbHB6zIR/bfn/4I1f3xz793B37benr6+Mq/8Su87g+kNzekb6/6APvrPn98l/g7w8+/8Wfqfvox4e/PiT+Hf2zo68t3HCJ/R/v6yi9mvbEhPb8LfbD7/vLR3f9r6v3L14Czn6Tf29/88xsN3wIvf3/5BP+6TzxvbEgfIuX+KfBlpN9+XvXa13aeYl+e/f7yFdrX/bw3NaRv7z8Evt39H/7Y//lOe/B/Cdv/dO77HL/w3bO3NKQvqS/Y/fl9pMBnjeSQ/h3a/3fS33/+Ofwtcu1Xf9H/DQ0p0+yLn/9kw/PH1DdWMnV/evky5fNr/8rwK378jvX88sWGP+wvv/xp+MrvYb2hIf2W/N059JXgnzJ1P/8ZOvFF/99zdb9/3Re/39SQkn9f+PTu6X3qG/2pT9LPudBfPmS+FfwLXfPfRwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECDAkQIAhAQIMCRBgSIAAQwIEGBIgwJAAAYYECPw/nuSxy9P89QIAAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title \"ir1\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(ir1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 5 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         2         2\n",
       "  [2]         3         3         1\n",
       "  [3]         4         4         1\n",
       "  [4]         5         5         1\n",
       "  [5]         6         7         2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disjoin(ir1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAWpElEQVR4nO3d2UJdV5JF0UsnS1jA//9t0QmwrZ1lofA6sNcYD6U0ak7OUISAi1x5ugN+2+no/wKwA4cEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHFLW6XR6883PvutfvfXu++l0ufjey+e33X/7/Tf+m/JLHFLW1CE9X8lPvvf6x9ueb40Ih5Q1dEj3R3L+8+99uKPnt517l5TjkLIW72B+1ZfT6dvP3n5zeXo9pG+n05eJh/EvOKSQm6uz0/nXv71Huv3j4v4/XT7dxMsBXF/d/8er67u3b3345vr+B189vY85e/6xr997c/5wNff/4eL1Vk+ns0Qadw4p5c/H9xQvW/70zc3Z6fnNr2+7u7t4fuPTJzivP+HL05u/P/1qf//e88df5eHtr4d0/+7pz1xiN4eU8eNk/nJIV48foN3eX87Xl7fdXb78wMvXH3n3+tOv7v/p69PP+Nv33v9aF9dvP3p8+WH85xxSxP2nK2f3O3599pdDeviI7P6b26dXDp7e9vBiwdfb+w/67r+9vnt7Kg+/wNXTP768sv36vRe3P571ekhet8txSBGXT1fx8tr00zdnbz4XevNe6um9yJendz6vP+HhB94+/eP50wX+/Xvf/joPbp5f2+O/55AiXpb7L4f0x/MHa9dvvuv+/z69b7l5+ee3v8Bf3vj2e1/eIb19YXDoNUL+fwYd8fNDunt+/eB0dnP3jxu5+7VD+smzHFKQQUesNv/228XLy3b/fI909s+f4JA+KIOO+PE50refbP711dvruFx+jnT3+q1D+nAMOuLr04tu3/76qt3583uft+971q/a3b1+e/HPFxten/WXFxu8ahfikDJ++nWk+6O5uHl8zeHLy9tevh779A7p54d09eMrrf/7kP58/jX47zmkjOe/2XD58xcbfvo3G55u4KeH9JMvyL4+6vUffEE2xyGF3Ny/F7n4+9+1e/r86OIvR/HwxrOf/l2712+///OvCL0+6fUf/BtJOQ7pw/iVVwbO/tWP9ZdWcxzSR3H9K4f05e1fZfgfv6J/jSLFIX0Mj58t/fuX2G6eP6/6ny5/vLbHf88hfQyPf+f7/38n8/bH/3+f/vgrq0kO6WO4P4zzn/47rwv/4kq81JDkkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQYUHVIp+OI3lxXbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbndcU27lRldF5XbONOVUbnVcXCf8UhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMWB7SaUfJyX6YaVZG/2eWsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZawoJonexjDWFBNG7WMaaQoLoXSxjTSFB9C6WsaaQIHoXy1hTSBC9i2WsKSSI3sUy1hQSRO9iGWsKCaJ3sYw1hQTRu1jGmkKC6F0sY00hQfQulrGmkCB6F8tYU0gQvYtlrCkkiN7FMtYUEkTvYhlrCgmid7GMNYUE0btYxppCguhdLGNNIUH0LpaxppAgehfLWFNIEL2LZawpJIjexTLWFBJE72IZm5ws7MohwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgw4H2HdIJtJQ/pXT9rgieXPPqzPdkhefKHfPRne7JD8uQP+ejP9mSH5Mkf8tGf7ckOyZM/5KM/25Mdkid/yEd/tic7JE/+kI/+bE92SJ78IR/92Z7skDz5Qz76sz3ZIXnyh3z0Z3uyQ/LkD/noz/Zkh+TJH/LRn+3J/vY3DHBIMMAhwQCHBAMcEgxwSDDAIcEAhwQDHBIMcEgwwCHBAIcEAxwSDHBIMMAhwQCHBAMcEgx41yF9Pej8vp6fzr7cHvDg26vT6er7AQ9+9Ocx4/6d/22G3/T9Yd43+ef+xv8gxXvm9P2Y6d59eWw8O+CSzh6ffNAl3Z4ds8zHHdL1Ub/RP+7o7B0/9dd/yvezg/6YOl3dPrw3vIo/+cvDM7+cLuMPfnR51LgP6r3/c+vs+93t5enLQY+/Pv356z/p13+Pvp4ujvmdvXx66gEPPzvdHvPgB98O+vDq6+mPIx5791D8cEK373m/MOH27D1/gvz679F95UEr9fz8ox5+zG/szVF/bn09fT3isfeujvog+snl6T0fU/7679H3A3f57uEPqotjHvzlmMW6ON0c9QHA9dXp7IgPr85Pd3+cPX4cf4Tv7/uQ8l2/R0ce0tfT9RGPvf8A65AP2f84fTto3JdPn3cf8MfW6XT5zs/4J7zvHdKnO6Sbd30A+/u+Xp4d8TnD42f8x4z7dH/Cd7dHvB++v6HvD19xOORztO/vfDHrkx3S7dlBH9jdPXzont+p84fXgI/9SPo8/synLzTcHPDku4eP4N/3Ec8nO6SLQ4b75ICXka4ef1fbXts5Hfby7N3DK7Tv+3mf6pBuzi8O+HL3i3z2y1faD/z/hJ1/9HFf5/iNr559pkO6PuoFu6evIx3wscaRh/QjOv856R+P74dvDvndfveL/p/okI6Z7IPHv9lwe3nUF1aOGfeXh5cpb9/7KcPvuP8T6/bhxYZv8Sc/vDd859ewPtEhXR35p/NBrwQ/Ombct0/RR7zo/8dx4z5/34vfn+qQjvx84cvZ6fyoL/Qf9ZH07XHR1xfHfCn4N2bt30eCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYIBDggEOCQY4JBjgkGCAQ4IBDgkGOCQY4JBggEOCAQ4JBjgkGOCQYMD/ATOPT9Jnd8VJAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"disjoin(ir1)\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(disjoin(ir1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manipulating IRanges, intra-range\n",
    "“Intra-range” manipulations are manipulations where each original range gets mapped to a new range.\n",
    "\n",
    "Examples of these are: `shift()`, `narrow()`, `flank()`, `resize()`, `restrict()`.\n",
    "\n",
    "For example, `resize()` can be extremely useful. It has a fix argument controlling where the resizing occurs from. Use fix=\"center\" to resize around the center of the ranges; I use this a lot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 4 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         1         1\n",
       "  [2]         3         3         1\n",
       "  [3]         7         7         1\n",
       "  [4]         9         9         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resize(ir, width = 1, fix = \"start\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAbb0lEQVR4nO2diVIbhxYFJXYwy///7UMLILEE4XdzznTorood40G0jqa1QSqrJxH5v1m1BUT+CxiSyACGJDKAIYkMYEgiAxiSyACGJDKAIYkMYEgiAxiSyACG9JHV6ptV7leri1OO++aCX/94edqXPebii8Ov16vV5Y8vbXf4p590+Y+fePnus5+17n/yhf8rGNJHvj0H9yfLUEh/1qvTvuwRd18cfv388cmQXuw+Z/+3B5+9v5P5dRjSR747B5/PlbNTjvvugvd/fPnojy5u09Gnh5/tHxGmQvrnyzm6Crtfz37nQ5Ih/Zyr1ep24nL+PqSHi9VXIf247h1nq/Xrrz+4vP3fHn727Wp19TcGcH5dSM83/cPZ9qZ+vFqv1lcP248+Xp8//8XF7cshu1/3PL07eH185r9d4gHP98t/Np/2/JePz78/bB/EXk7Kh8v16uxm98e3r7H59c+zxuXDse6Rx+tHz78Jafv7w8vzrPOdztcXebE9cPfrwRgHh91u4j3bTbC/yq9/e/jZz3/5ocZfwG8M6fkkP38+yda702Bzgr38++bjn4Z0ePCfl7PzLaSz/We+cb27X948A9uckM9309dvL4r2X+tjSLtTc31Q0lchXXz+UHGc5dPTzfbrbh5DL98fc3yR11vd7a+HY7wddv7yKX/ervLr3x589vYV5J+nX8dvDGl3cr+cLpu7z8vtRx6fz4ybp09DOjz4ZnfUYUirD0/2dg9BmwvensLPl/zw+gnrw4s+OvX3HLxP9nlI53dfPOd6H9LmC99tMl5/OObdRb5xOMbrYc9X+vxxW+TF21X+/EJe9/lV/MaQNmfE/sx43J01q+15vnkidvZ0fIpebU7E44Nf3+B9C2l7icecbZ/TvZ3Wbxe8Oa2fL/Nu/fb5Lxe3/rP9iic9Nfoig4NL2/z2HPT6ZvWjx4hPxzjbffDtTmZ3lT+T+J3v2/3GkO42v1/sXrxsnyVtHyMu7w4OefnXXUfHB7+cUwch3T194HoT3Z/9k6G73TOs3Sdc7I+/+xjS5uOPXxTyyRX5pw+//H7z/jHuW74c4+2Pr1f5M4n9g/Ev4zeG9HJf+vZ07Xr/lOr92bHv6Pjg179+C+HjA9LmdLrcvr+3ebF0tWvv6EnXl+/aHZ2cXz8POzGkTffHD3HfPbX7aoyH26vz1wfY/VX+9CJOvCf4b/HrrvLBiXZwNr28AbV+ODhkc29+9fHgr0J4x+bN4M0/6+f75/XuVK6EtOni+t0R/xjS52Pcnh18zvurcJLZf5tfd5VfbuX18a39eLt7D+r84JC71++IrN+d20f/8sV5s3k02z8avVxQI6THzTO19ePxEf8c0mdjbN5YOLu8uTekL/h1V/nlVr74+Mrm7vLoNHnr6PjgE0PavT66e32d9Hrky4XdnhLS91fkiw8fXM/Vz14j7Xg3xtnbk9wnQ/qEX3eVX5+q7N4ju93e7569vnBavx5y0NHxwecf32z49Cut354H7V+k7P58s3vX7vbwXbvHp38ppOeM1/frH71r93GMx9eL++wR6f0LxAfftfsNvJ4Dr9/N+bON5vzh9buob+f/6/Ofw4MvX07LLx5RXrjaPxJs7tqvDo84+j7S7o9X/3A531yRd4e/+/B681793Y/eRzseY293vv315S37oxHf/0DQn795AMTze0O625/L2/Pg5fX1Vz/ZcHjwJ9+QPb7kPZvndLf7zz0qb/+TDRf7P16ujl+aTYZ0vXO9fvd+wz9zNMbebu+82j24vX7NF/dD/Ibsr+DtvHu8en4Sc7F/7bN9SXB+GMjxw8bBwfcff0To3SXvWb+90378tR8ut1/r5Y8Xu7vwfyGk12dZFy9PR0/icIwXu/vnj60v73eX+PY1Lz4+/PzO/yLp14U0wPqLM730GntpL+39oVU5jatPf5Th6bHzGrv0Zb/kzv+MQk7j4ePrgg03nR96Ln3ZL/nZs8j/DIb0F3z+KuCic0KXvuxX/M4fWTWkv+K3niyn8DvfajAkkREMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBjAkkQEMSWQAQxIZwJBEBvgypNUg/6L+v8m/p00FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSkBraVKBzJywNqaFNBTp3wtKQGtpUoHMnLA2poU0FOnfC0pAa2lSgcycsDamhTQU6d8LSs0VkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAYwJJEBDElkAEMSGcCQRAZYXEgrke9pn6YfWJzR4oROBOqt9hCLM1qc0IlAvdUeYnFGixM6Eai32kMszmhxQicC9VZ7iMUZLU7oRKDeag+xOKPFCZ0I1FvtIRZntDihE4F6qz3E4owWJ3QiUG+1h1ic0eKETgTqrfYQizNanNCJQL3VHmJxRosTOhGot9pDLM5ocUInAvVWe4jlGYkAMSSRAQxJZABDEhnAkEQGMCSRAQxJZABDEhnAkEQGMCSRAQxJZABDEhnAkEQGMCSRAQxJZABDEhnAkEQGWFZIN2er9dVj2+Kv+LOsJU/i/nK1unxoW/yUx6v1As+SRd38V9v/Y8d6aRudwuN6UUuexB1y7of1TnthdwBLuvnvV5fPN+rN6rIt8hdcLPB/2fMd6/X90+PF6qrt8TMut8JXSztLlnTzX+xkgKfk0+0S/99X33C7PSMfV+u2yM9YLfMsWZjOhqVNdAIPq3Oe9eXqvq3wN+yfQy+t/+Xd/I+r87bCjzlfPfBCOls9Xa+3z6ZRXO+f2l23RY5Z3s1/s7prK/yU69Ut8HF0tbrYvmpve/yUm827DeubtsY7FnfzP6wv2go/5X51QXxC+nw23j89Xi7trv1brrfv2i3Nemk3/+Oa98TubPMOMjGkzWukh9VZW+Rn3Gye2j33v7CHpKXd/Oewm/Vp86J981yUGNLhbxjOVptXdY9L639ZKz6cnS/s+2wnsHqlbfIzoN9tWGj/i9K5A75hxw3pevtI+kDbfPf29+K+/bWkGx93mx5Cy2j76uhx82Ljti3yM65Wm5+zu1raD2Qs6ea/ZN6z7wBa797+wt15nS9Se0k3P/Qp0g6i9d35ar2wO/ZT2P70d1viPcCbX2R5GJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQxgSCIDGJLIAIYkMoAhiQzwP51CKKYSHDKoAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"resize(ir, width = 1, fix = \"start\")\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(resize(ir, width = 1, fix = \"start\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 4 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         2         2         1\n",
       "  [2]         3         3         1\n",
       "  [3]         7         7         1\n",
       "  [4]         9         9         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resize(ir, width = 1, fix = \"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAMFBMVEUAAABNTU1oaGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////QFLu4AAAACXBIWXMAABJ0AAASdAHeZh94AAAXfElEQVR4nO3di1ZiWbKGUVC8K77/2xabSwiKmUL9xNJ0zjFOVVcWyQ5c8QFi9unZK/C/zUYPAP8CIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQdmazv3wtnmezxVdu95c7rn+8+dplDy0+ufndfDa7OfneNjc/+RHdfOkeV6M+n3a/P5mQdv66TtvFCIX0NJ997bIHHj+5+d3q17tC2k3+t3vcPvH8EkLa+ds6rfbi6iu3+9sdb/9x96sn3d3U0dGbX22f/TtC+uPN9+/x6je9JAnpq25ns4fE/Zwf0sti9llIJ9e9cTWb11+/7I/X2r/Hh9ns9pypfqRfEtLq8F+u1se6vJ3P5rcv619d3l2v/sXiYXeTzV+3Xt/deH64+W/3uGf1HPw0/bbVv1yu/v6yfhHbLd7LzXx2db/5x7drTH99Wo1x83I47sEc9avXfwlp/feX3Xuq6804n9/lYn3DzV9XD3Y1/vXj5t/sP/LpdzxOIz6/7k/+/jbrL8fBPc5OK/RH+z0hTVuyWrL5ZhGmBdv95+nXj4a0f+On3Xa+hXS1/Z1v7jbPwdM7sKnO1VPy3ds3RdtrfQzpdv2f53slfRbS4vjLwWGWr6/36+tOr6E3729zeJd363E3f9092PVzw/4jfxtxeqd2/KtTX479e5y+q6yQ/3W/J6TNcu/amZ4qb9a/slyt9v3r0ZD2b3y/udV+SLMPb/Y2L0HTHa9XeHXPL/Ub5vt3fbD6W3ufhR0PaXq1+FJI04Ufp4znH27z7i731HyP7x754YjHvzrHvxz1NfsFfk9I19Obrfv135ebhGbrPZ/eiF29Hq7o7XqdDm5cH+a+hbS+x0NX6/d0b2v9dsfTWq/u83H+9vt3dzd/Wl/xS2+DPslg796mv62Cnt/PTno9WN18/rx+Vrl698jXIz5unh72rvHhNh+/HL/pc7vfE9L6zf9i883L+l3S+in15nHvJrv/uOno8MZXm+r2Q3p8/eBuWqvNm7in6R3eXf2Gxfb2jx9Dmn59+dkLxccH8qdf3v39/v1r3F9t51te3b28e+TvR9z+7ehtDm1foH+F3xPS9tR35tsfvlRLbyu67ejwxvWv39bp4wvStDo368/3pu81bjftHbzp+vRTu4NCPn8f9sWQpu4PX+I+v8sjd3zwyN/f+dvfDm9z5MvxxWeHf8EveaB7u7C3Tbe7TXjZu8n0bH778cafhfDO9MHv9H/z1XPxvL59eG0PaXqSuHt3i3NC2h/9s5A+/1mUkP41uxOdH57s8uF6vQnXezd5rJ9+zN/t1sF/+GRHplez7avR7o5GhLSc3rbOl4e3+HpIh1+m4yEdv82Xpv0X/ZIHujvRxce38o83B5v+1tHhjb8Y0ub7o8f6Pqluubuzh6+E9PcH8skv7z3O2WnfI12/+x5p78t0PKTjt/nStP+iX/JAdyf6sPmM7GH9InRV3zi9fSew19Hhja8/fthw9Erzt/c6229SNv98v/no62H/U7vl64VCepo+gpuf/6ndwSM/FtLys9scePGp3b+mzrl+WvK0jub6pX6K+rb/9f5n/8Y3u7X85BVl53b7SjC9zt3u3+Lg50ibf7z9w/385YG8u/m7X55Pn9U/nvaZWc13/+6RH7nz9eRHb3Pg6bQXxR/tt4X0uD389Y7vPmz47E827N/4yA9kD+95a3pP97D9vQflbf9kw2L7jzezo0/3X30gfwzpbjPr3bvPG/7saf9PNhx8md5dczf50dsc8APZf87bOa//RNli++5+/f3R9X4ghy8bezd+/vhHhN7d89b87ZP2w2u/3KyvtfvHxebp+gIh1Tuqxe7t6JdMf3SuvjL7X6b319xOfvQ2B37TfyPpl4QUMP9k0wd9P/0Tvo33h1b56PboH2V4XY75fnrQZU/y6L9GwUcvH/6s99r9mD/gPOiyJzntneUPJ6QvO/6OfzFmoQdd9hS/6Y+sCukEv2sx/r/f9FGDkCBCSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgT0hTRr0/aQYEdIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgTYOggQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIEPBpSLO4zofFOfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omFxIlf+ajd6FvciFR8mc+ehf6JhcSJX/mo3ehb3IhUfJnPnoX+iYXEiV/5qN3oW9yIVHyZz56F/omt94QICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBBwXkgz+Gd1hnTW7/rGF/rnHtC/d6Fv/oCE1HodF/r21xHST7iOC3376wjpJ1zHhb79dYT0E67jQt/+OkL6CddxoW9/HSH9hOu40Le/jpB+wnVc6NtfR0g/4Tou9O2vI6SfcB0X+vbXEdJPuI4LffvrCOknXMeFvv11vnsS8C8TEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIOCOk+6vZ/HaZH+W95c1sdvN8+etMnjqeUP7P/9jBaZ6nL93LxS/z//73G06wvJ337Nzr6kLXj+f8xtO/BrfrL9388o9qvr5QS0nLecN6P7eF9Nh0RLuO5he+zstmFeaXf2q4Xl/o7ozfefKxPs9uVgd0P7s542InuZ0ucTtbXPo6k0XHej/3PJaV+fz5dbmY3fZc7XH2dOEr3Kwfyu3ld+5+dr2c3gmd8ex98gItNr/j8ps3ny1brrPy0PI6cX/WM90ZHtZ7t7z4C8XGcn7x54dZ185dr58TXs55Cjp3tpYFn67TsA0vs+uekO4vf5HJWc+oZ1vMLv4WcvvG+/K7sCv2+ozfet4Vl+dc6wy3Hct3PXvpCGkxe7xZfct8+QtdzV7v5ut34A2eG95B3m3f2l38Ff1/vPSduUD3s7M+2jjR6i1Xw97dzR5aXmAXm+/ML/8MNJstOj4C2Gh4QVqt2/Rpw/zyz6lXs+nzjKe+kF4u/754cr+YX/5ZaP0ZQEdIs1Wwr8uG19jVzj1P3zJ3fEv2fPlPAF6n57pzP0w79TqL5evzWW/0z1qg5bznjd3r9H7/0nt3NX1M3PUt3/Sm+OrSl9j8zODl8hd6nd5vNbw1uZ/emCwvvwrbH7mc9RnuWQt03XFEGxf/7OlmvQl9ITVcqu1Drtf6HOCyrtZvHxuegqZa53dd3yO9XF1f/idj5dLrUD+db/v/YX3xC7X9hKLrZ2Odzwyv04M6o9jTZ3ts+sBu83Oki79B6Qtp94Auvnp36xfZl45z6vlIf/Oy1/CDsc0R3Z9zRCevT8v5TNY/yF4uen740vFkd7t+p9/wTcXqyWf98/mHS19oevHr+JHV6iu33H79Ln2h1c49XZ3zlTt5gW7a3gnNmz4tnnSEtNw8oJYP9Lu+clcdH37v/gjc5R/Q9ojOec9w8gI1fktxO59dNf1hgJa338u2B/R43fKT375vW9Z/+rvhOi+r14lF05/+Bj4QEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAH/AetVc23W+01wAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"resize(ir, width = 1, fix = \"center\")\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotRanges(resize(ir, width = 1, fix = \"center\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manipulating IRanges, as sets\n",
    "\n",
    "Manipulating IRanges as sets means that we view each IRanges as a set of integers; individual integers is either contained in one or more ranges or they are not. This is equivalent to calling reduce() on the IRanges first.\n",
    "\n",
    "Once this is done, we can use standard: `union()`, `intersect()`, `setdiff()`, `gaps()` between two IRanges (which all returns normalized IRanges)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 2 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         1         1\n",
       "  [2]         3         6         4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir1 <- IRanges(start = c(1, 3, 5), width = 1)\n",
    "ir2 <- IRanges(start = c(4, 5, 6), width = 1)\n",
    "union(ir1, ir2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 1 range and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         5         5         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "intersect(ir1, ir2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 2 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         1         1\n",
       "  [2]         3         6         4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Because they return normalized IRanges, an alternative to union() is\n",
    "reduce(c(ir1, ir2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is also an element-wise (pair-wise) version of these: punion(), pintersect(), psetdiff(), pgap(); this is similar to say pmax from base R. In my experience, these functions are seldom used."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding Overlaps\n",
    "Finding (pairwise) overlaps between two IRanges is done by `findOverlaps()`. This function is very important and amazingly fast!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 3 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         2           1\n",
       "  [3]         2           2\n",
       "  -------\n",
       "  queryLength: 3 / subjectLength: 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir1 <- IRanges(start = c(1,4,8), end = c(3,7,10))\n",
    "ir2 <- IRanges(start = c(3,4), width = 3)\n",
    "ov <- findOverlaps(ir1, ir2)\n",
    "ov"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It returns a Hits object which describes the relationship between the two IRanges. This object is basically a two-column matrix of indicies into the two IRanges.\n",
    "\n",
    "The two columns of the hits object can be accessed by `queryHits()` and `subjectHits()` (often used with unique()).\n",
    "\n",
    "For example, the first row of the matrix describes that the first range of ir1 overlaps with the first range of ir2. Or said differently, they have a non-empty intersection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 0 ranges and 0 metadata columns:\n",
       "       start       end     width\n",
       "   <integer> <integer> <integer>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "intersect(ir1[subjectHits(ov)[1]],\n",
    "          ir2[queryHits(ov)[2]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The elements of `unique(queryHits)` gives you the indices of the query ranges which actually had an overlap; you need unique because a query range may overlap multiple subject range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 2 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "queryHits(ov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(queryHits(ov))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*The list of arguments to `findOverlaps()` is long; there are a few hidden treasures here. For example, you can ask to only get an overlap if two ranges overlap by a certain number of bases.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (query, subject, maxgap = -1L, minoverlap = 0L, type = c(\"any\", \n",
       "<span style=white-space:pre-wrap>    \"start\", \"end\", \"within\", \"equal\"), select = c(\"all\", \"first\", </span>\n",
       "<span style=white-space:pre-wrap>    \"last\", \"arbitrary\"), ...) </span>\n",
       "NULL</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (query, subject, maxgap = -1L, minoverlap = 0L, type = c(\"any\", \n",
       "    \"start\", \"end\", \"within\", \"equal\"), select = c(\"all\", \"first\", \n",
       "    \"last\", \"arbitrary\"), ...) \n",
       "NULL\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (query, subject, maxgap = -1L, minoverlap = 0L, type = c(\"any\", \n",
       "    \"start\", \"end\", \"within\", \"equal\"), select = c(\"all\", \"first\", \n",
       "    \"last\", \"arbitrary\"), ...) \n",
       "NULL\n",
       "```"
      ],
      "text/plain": [
       "function (query, subject, maxgap = -1L, minoverlap = 0L, type = c(\"any\", \n",
       "    \"start\", \"end\", \"within\", \"equal\"), select = c(\"all\", \"first\", \n",
       "    \"last\", \"arbitrary\"), ...) \n",
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "args(findOverlaps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### countOverlaps\n",
    "For efficiency, there is also countOverlaps(), which just returns the number of overlaps. This function is faster and takes up less memory because it does not have to keep track of which ranges overlap, just the number of overlaps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>0</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 0\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 0\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 2 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "countOverlaps(ir1, ir2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding nearest IRanges\n",
    "Sometimes you have two sets of IRanges and you need to know which ones are closest to each other. Functions for this include `nearest()`, `precede()`, `follow()`. Watch out for ties!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 3 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         3         3\n",
       "  [2]         4         7         4\n",
       "  [3]         8        10         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 2 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         3         5         3\n",
       "  [2]         4         6         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "3. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 1 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nearest(ir1, ir2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 02: GenomicRanges - GRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install Genomic Ranges Library\n",
    "library(GenomicRanges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**`GRanges`** : `GRanges` are like `IRanges` with strand and chromosome. Strand can be +, - and *. The value * indicates ‘unknown strand’ or ‘unstranded’. This value usually gets treated as a third strand, which is sometimes confusing to users (examples below)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-3      +\n",
       "  [2]     chr1       3-5      -\n",
       "  [3]     chr1       5-7      +\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## They get created with the GRanges constructor:\n",
    "gr <- GRanges(seqnames = \"chr1\", strand = c(\"+\", \"-\", \"+\"),\n",
    "              ranges = IRanges(start = c(1,3,5), width = 3))\n",
    "gr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Natural accessor functions: \n",
    "- strand(), \n",
    "- seqnames(), \n",
    "- ranges(), \n",
    "- start(), \n",
    "- end(), \n",
    "- width().\n",
    "Because the have strand, we now have operations which are relative to the direction of transcription (`upstream()`, `downstream()`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       4-5      +\n",
       "  [2]     chr1       1-2      -\n",
       "  [3]     chr1       8-9      +\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "flank(gr, 2, start = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seqinfo object with 1 sequence from an unspecified genome; no seqlengths:\n",
       "  seqnames seqlengths isCircular genome\n",
       "  chr1             NA         NA   <NA>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# GRanges, seqinfo\n",
    "# GRanges operate within a universe of sequences (chromosomes/contigs) and their lengths.\n",
    "# This is described through seqinfo:\n",
    "seqinfo(gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seqinfo object with 1 sequence from an unspecified genome:\n",
       "  seqnames seqlengths isCircular genome\n",
       "  chr1             10         NA   <NA>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlengths(gr) <- c(\"chr1\" = 10)\n",
    "seqinfo(gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'chr1'"
      ],
      "text/latex": [
       "'chr1'"
      ],
      "text/markdown": [
       "'chr1'"
      ],
      "text/plain": [
       "[1] \"chr1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlevels(gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>chr1:</strong> 10"
      ],
      "text/latex": [
       "\\textbf{chr1:} 10"
      ],
      "text/markdown": [
       "**chr1:** 10"
      ],
      "text/plain": [
       "chr1 \n",
       "  10 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlengths(gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1         4      +\n",
       "  [2]     chr1      8-10      +\n",
       "  [3]     chr1       1-2      -\n",
       "  [4]     chr1      6-10      -\n",
       "  [5]     chr1      1-10      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Especially the length of the chromosomes are used by some functions. For example gaps() return the stretches of the genome not covered by the GRanges.\n",
    "gaps(gr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example, we know that the last gap stops at 10, because that is the length of the chromosome. Note how a range on the * strand appears in the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let us expand the GRanges with another chromosome\n",
    "seqlevels(gr) <- c(\"chr1\", \"chr2\")\n",
    "seqnames(gr) <- c(\"chr1\", \"chr2\", \"chr1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-3      +\n",
       "  [2]     chr2       3-5      -\n",
       "  [3]     chr1       5-7      +\n",
       "  -------\n",
       "  seqinfo: 2 sequences from an unspecified genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you `sort()` a GRanges, the sorting order of the chromosomes is determined by their order in seqlevel. This is nice if you want the sorting “chr1”, “chr2”, …, “chr10”, …"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-3      +\n",
       "  [2]     chr1       5-7      +\n",
       "  [3]     chr2       3-5      -\n",
       "  -------\n",
       "  seqinfo: 2 sequences from an unspecified genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-3      +\n",
       "  [2]     chr2       3-5      -\n",
       "  [3]     chr1       5-7      +\n",
       "  -------\n",
       "  seqinfo: 2 sequences from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# You can associate a genome with a GRanges.\n",
    "genome(gr) <- \"hg19\"\n",
    "gr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This becomes valuable when you deal with data from different genome versions (as we all do), because it allows R to throw an error when you compare two `GRanges` from **different genomes**, like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in mergeNamedAtomicVectors(genome(x), genome(y), what = c(\"sequence\", : sequences chr1, chr2 have incompatible genomes:\n  - in 'x': hg19, hg19\n  - in 'y': hg18, hg18\n",
     "output_type": "error",
     "traceback": [
      "Error in mergeNamedAtomicVectors(genome(x), genome(y), what = c(\"sequence\", : sequences chr1, chr2 have incompatible genomes:\n  - in 'x': hg19, hg19\n  - in 'y': hg18, hg18\nTraceback:\n",
      "1. findOverlaps(gr, gr2)",
      "2. findOverlaps(gr, gr2)",
      "3. .local(query, subject, maxgap, minoverlap, type, select, ...)",
      "4. findOverlaps_GNCList(query, subject, maxgap = maxgap, minoverlap = minoverlap, \n .     type = type, select = select, ignore.strand = ignore.strand)",
      "5. merge(seqinfo(query), seqinfo(subject))",
      "6. merge(seqinfo(query), seqinfo(subject))",
      "7. .Seqinfo.merge(x, y, ...)",
      "8. .Seqinfo.mergexy(x, y)",
      "9. mergeNamedAtomicVectors(genome(x), genome(y), what = c(\"sequence\", \n .     \"genomes\"))",
      "10. stop(msg)"
     ]
    }
   ],
   "source": [
    "gr2 <- gr\n",
    "genome(gr2) <- \"hg18\"\n",
    "findOverlaps(gr, gr2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenomicRanges - Basic GRanges Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DataFrame\n",
    "The [S4Vectors](https://bioconductor.org/packages/release/bioc/html/S4Vectors.html) package introduced the DataFrame class. This class is very similar to the base data.frame class from R, but it allows columns of any class, provided a number of required methods are supported. For example, DataFrame can have IRanges as columns, unlike data.frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 2 rows and 1 column\n",
       "    iranges\n",
       "  <IRanges>\n",
       "1       1-3\n",
       "2       2-4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ir <- IRanges(start = 1:2, width = 3)\n",
    "df1 <- DataFrame(iranges = ir)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges object with 2 ranges and 0 metadata columns:\n",
       "          start       end     width\n",
       "      <integer> <integer> <integer>\n",
       "  [1]         1         3         3\n",
       "  [2]         2         4         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df1$iranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>iranges.start</th><th scope=col>iranges.end</th><th scope=col>iranges.width</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>3</td><td>3</td></tr>\n",
       "\t<tr><td>2</td><td>4</td><td>3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " iranges.start & iranges.end & iranges.width\\\\\n",
       " <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 1 & 3 & 3\\\\\n",
       "\t 2 & 4 & 3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| iranges.start &lt;int&gt; | iranges.end &lt;int&gt; | iranges.width &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 1 | 3 | 3 |\n",
       "| 2 | 4 | 3 |\n",
       "\n"
      ],
      "text/plain": [
       "  iranges.start iranges.end iranges.width\n",
       "1 1             3           3            \n",
       "2 2             4           3            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df2 <- data.frame(iranges = ir)\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the data.frame case, the IRanges gives rise to 4 columns, whereas it is a single column when a DataFrame is used."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRanges, metadata\n",
    "`GRanges` (unlike IRanges) may have **associated metadata**. This is immensely useful. The formal way to access and set this metadata is through values or elementMetadata or mcols, like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 1 metadata column:\n",
       "      seqnames    ranges strand |     score\n",
       "         <Rle> <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1       1-3      + |       0.1\n",
       "  [2]     chr1       3-5      - |       0.5\n",
       "  [3]     chr1       5-7      + |       0.3\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr <- GRanges(seqnames = \"chr1\", strand = c(\"+\", \"-\", \"+\"),\n",
    "              ranges = IRanges(start = c(1,3,5), width = 3))\n",
    "values(gr) <- DataFrame(score = c(0.1, 0.5, 0.3))\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0.1</li><li>0.5</li><li>0.3</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.1\n",
       "\\item 0.5\n",
       "\\item 0.3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.1\n",
       "2. 0.5\n",
       "3. 0.3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.1 0.5 0.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# A much easier way to set and access metadata is through the $ operator\n",
    "gr$score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 2 metadata columns:\n",
       "      seqnames    ranges strand |     score    score2\n",
       "         <Rle> <IRanges>  <Rle> | <numeric> <numeric>\n",
       "  [1]     chr1       1-3      + |       0.1      0.02\n",
       "  [2]     chr1       3-5      - |       0.5      0.10\n",
       "  [3]     chr1       5-7      + |       0.3      0.06\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr$score2 = gr$score * 0.2\n",
    "gr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### findOverlaps\n",
    "`findOverlaps` works exactly as for IRanges. But the strand information can be confusing. Let us make an example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-3      *\n",
       "  [2]     chr2       3-5      *\n",
       "  [3]     chr1       5-7      *\n",
       "  -------\n",
       "  seqinfo: 2 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr2 <- GRanges(seqnames = c(\"chr1\", \"chr2\", \"chr1\"), strand = \"*\",\n",
    "               ranges = IRanges(start = c(1, 3, 5), width = 3))\n",
    "gr2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 2 metadata columns:\n",
       "      seqnames    ranges strand |     score    score2\n",
       "         <Rle> <IRanges>  <Rle> | <numeric> <numeric>\n",
       "  [1]     chr1       1-3      + |       0.1      0.02\n",
       "  [2]     chr1       3-5      - |       0.5      0.10\n",
       "  [3]     chr1       5-7      + |       0.3      0.06\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note how the ranges in the two GRanges object are the same coordinates, they just have different seqnames and strand. Let us try to do a standard `findOverlaps`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 4 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         2           1\n",
       "  [3]         2           3\n",
       "  [4]         3           3\n",
       "  -------\n",
       "  queryLength: 3 / subjectLength: 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "findOverlaps(gr, gr2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice how the `*` strand overlaps both + and -. There is an argument `ignore.strand` to findOverlaps which will … ignore the strand information (so + overlaps -). Several other functions in GenomicRanges have an ignore.strand argument as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 4 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         2           1\n",
       "  [3]         2           3\n",
       "  [4]         3           3\n",
       "  -------\n",
       "  queryLength: 3 / subjectLength: 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "findOverlaps(gr, gr2,ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### subsetByOverlaps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A common operation is to select only certain ranges from a GRanges which overlap something else. Enter the convenience function `subsetByOverlaps`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 2 metadata columns:\n",
       "      seqnames    ranges strand |     score    score2\n",
       "         <Rle> <IRanges>  <Rle> | <numeric> <numeric>\n",
       "  [1]     chr1       1-3      + |       0.1      0.02\n",
       "  [2]     chr1       3-5      - |       0.5      0.10\n",
       "  [3]     chr1       5-7      + |       0.3      0.06\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "subsetByOverlaps(gr, gr2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### makeGRangesFromDataFrame\n",
    "A common situation is that you have data which looks like a GRanges but is really stored as a classic data.frame, with chr, start etc. The makeGRangesFromDataFrame converts this data.frame into a GRanges. An argument tells you whether you want to keep any additional column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-4      *\n",
       "  [2]     chr1       2-5      *\n",
       "  [3]     chr1       3-6      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df <- data.frame(chr = \"chr1\", start = 1:3, end = 4:6, score = 7:9)\n",
    "makeGRangesFromDataFrame(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3 ranges and 1 metadata column:\n",
       "      seqnames    ranges strand |     score\n",
       "         <Rle> <IRanges>  <Rle> | <integer>\n",
       "  [1]     chr1       1-4      * |         7\n",
       "  [2]     chr1       2-5      * |         8\n",
       "  [3]     chr1       3-6      * |         9\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "makeGRangesFromDataFrame(df, keep.extra.columns = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenomicRanges - seqinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "## \n",
    "library(GenomeInfoDb)\n",
    "library(GenomicRanges)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop and keep seqlevels\n",
    "It is common to want to remove seqlevels from a GRanges object. Here are some equivalent methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1 range and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr2       2-5      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr <- GRanges(seqnames = c(\"chr1\", \"chr2\"),\n",
    "              ranges = IRanges(start = 1:2, end = 4:5))\n",
    "dropSeqlevels(gr, \"chr1\", pruning.mode=\"coarse\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1 range and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr2       2-5      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "keepSeqlevels(gr, \"chr2\", pruning.mode = \"coarse\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "gr <- GRanges(seqnames = c(\"chr1\", \"chrU345\"),\n",
    "              ranges = IRanges(start = 1:2, end = 4:5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1 range and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-4      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# You can also just get rid of weird looking chromosome names with keepStandardChromosomes().\n",
    "keepStandardChromosomes(gr,pruning.mode=\"coarse\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Changing style\n",
    "It is an inconvenient truth that different online resources uses different naming convention for chromosomes. This can even be different from organism to organism. For example, for the fruitfly (Drosophila Melanogaster) NCBI and Ensembl uses “2L” and UCSC uses “chr2L”. But NCBI and Ensembl differs on some contigs: NCBI uses “Un” and Ensembl used “U”."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1       1-2      *\n",
       "  [2]     chr1       2-3      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr <- GRanges(seqnames = \"chr1\", ranges = IRanges(start = 1:2, width = 2))\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr2      1-10      *\n",
       "  [2]     chr3      2-10      *\n",
       "  [3]     chrM      3-10      *\n",
       "  [4]     chr2      4-10      *\n",
       "  [5]     chr3      5-10      *\n",
       "  [6]     chrM      6-10      *\n",
       "  -------\n",
       "  seqinfo: 3 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr <- GRanges(rep(c(\"chr2\", \"chr3\", \"chrM\"), 2), IRanges(1:6, 10))\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'UCSC'"
      ],
      "text/latex": [
       "'UCSC'"
      ],
      "text/markdown": [
       "'UCSC'"
      ],
      "text/plain": [
       "[1] \"UCSC\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]        2      1-10      *\n",
       "  [2]        3      2-10      *\n",
       "  [3]       MT      3-10      *\n",
       "  [4]        2      4-10      *\n",
       "  [5]        3      5-10      *\n",
       "  [6]       MT      6-10      *\n",
       "  -------\n",
       "  seqinfo: 3 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlevelsStyle(gr)\n",
    "seqlevelsStyle(gr) <- \"NCBI\"\n",
    "gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'NCBI'</li><li>'Ensembl'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'NCBI'\n",
       "\\item 'Ensembl'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'NCBI'\n",
       "2. 'Ensembl'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"NCBI\"    \"Ensembl\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr2      1-10      *\n",
       "  [2]     chr3      2-10      *\n",
       "  [3]     chrM      3-10      *\n",
       "  [4]     chr2      4-10      *\n",
       "  [5]     chr3      5-10      *\n",
       "  [6]     chrM      6-10      *\n",
       "  -------\n",
       "  seqinfo: 3 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seqlevelsStyle(gr)\n",
    "seqlevelsStyle(gr) <- \"UCSC\"\n",
    "gr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "python3102jvsc74a57bd055db829a23e27f1a4129460ab65de434bf1fc9efa4ae0ce0afb72d96498ad818"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
