# Module 3 Exam Instructions **IMPORTANT**

For this project, it is recommended that you use the VMBox virtual environment provided with the Course package and the tools therein. You may also use your own system and software, however make sure that appropriate versions are installed. The answers are compatible with the following versions of the software: samtools v.1.2, bowtie v.2.2.5 and bcftools v.1.2. 

As part of the effort to catalog genetic variation in the plant Arabidopsis thaliana, you re-sequenced the genome of one strain (‘wu_0_A’; genome file: ‘wu_0.v7.fas’), to determine genetic variants in this organism. The sequencing reads produced are in the file ‘wu_0_A_wgs.fastq’. Using the tools bowtie2, samtools and bcftools, develop a pipeline for variant calling in this genome. NOTE: Genome and re-sequencing data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’. 

## Downlaod file
```{shell}
wget https://d396qusza40orc.cloudfront.net/gencommand/gencommand_proj3_data.tar.gz
```
## Apply to questions 1 - 5: 
Generate a bowtie2 index of the wu_0_A genome using bowtie2-build, with the prefix 'wu_0'. 

```{shell}
bowtie2-build wu_0.v7.fas wu_0
```

1. How many sequences were in the genome?
```
cat wu_0.v7.fas | grep ">" | wc -l
```
#### Ans: `7`

2.  What was the name of the third sequence in the genome file? Give the name only, without the “>” sign.
```
cat wu_0.v7.fas | grep ">"
```
#### Ans: `Chr3`

3. What was the name of the third sequence in the genome file? Give the name only, without the “>” sign.
```
cat wu_0.v7.fas | grep ">"
```
#### Ans: `mitochondria`

4.  How many index files did the operation create?
```
bowtie2-build wu_0.v7.fas wu-index/wu_0
```
#### Ans: `6`

5.  What is the 3-character extension for the index files created?
#### Ans: `bt2`



## Apply to questions 6 - 14: 
6.  How many reads were in the original fastq file?
```
cat wu_0_A_wgs.fastq | wc -l
```
#### Ans: `589416`

7. How many matches (alignments) were reported for the original (full-match) setting? Exclude lines in the file containing unmapped reads.
#### Ans: `147354 - 9635 = 137719`
```
bowtie2 -x index/wu_0 -U week03/wu_0_A_wgs.fastq -S week03/Arabidopsis_thaliana.sam
```
```
147354 reads; of these: 
  147354 (100.00%) were unpaired; of these:
    9636 (6.54%) aligned 0 times
    93780 (63.64%) aligned exactly 1 time
    43938 (29.82%) aligned >1 times
93.46% overall alignment rate
```

8. How many matches (alignments) were reported with the local-match setting? Exclude lines in the file containing unmapped reads. 
#### Ans: ` 147354 - 6310 = 141044`
```
bowtie2 -x index/wu_0 -U week03/wu_0_A_wgs.fastq -S week03/Arabidopsis_thaliana.local.sam --local
```
```
147354 reads; of these:
  147354 (100.00%) were unpaired; of these:
    6823 (4.63%) aligned 0 times
    88935 (60.35%) aligned exactly 1 time
    51596 (35.01%) aligned >1 times
95.37% overall alignment rate
```
9. How many reads were mapped in the scenario in Question 7?
#### Ans: `137719`

10. How many reads were mapped in the scenario in Question 8?
#### Ans: `141044`

11. How many reads had multiple matches in the scenario in Question 7? You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.
#### Ans: `43938`
```
147354 reads; of these: 
  147354 (100.00%) were unpaired; of these:
    9636 (6.54%) aligned 0 times
    93780 (63.64%) aligned exactly 1 time
    43938 (29.82%) aligned >1 times  <--- # Ans
93.46% overall alignment rate
```
12. How many reads had multiple matches in the scenario in Question 8? Use the format above. You can တ†nd this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.
#### Ans: `51596`

```
147354 reads; of these:
  147354 (100.00%) were unpaired; of these:
    6823 (4.63%) aligned 0 times
    88935 (60.35%) aligned exactly 1 time
    51596 (35.01%) aligned >1 times # Ans
95.37%
```
13. How many alignments contained insertions and/or deletions, in the scenario in Question 7?
#### Ans: `1429 + 1395 - 42 = 2782` 
````
cat Arabidopsis_thaliana.sam | grep -v "^#" | cut -f6 | grep 'I' | wc -l
cat Arabidopsis_thaliana.sam | grep -v "^#" | cut -f6 | grep 'I' | wc -l
cat Arabidopsis_thaliana.sam | grep -v "^#" | cut -f6 | grep 'I' | grep 'D'| wc -l
````

14. How many alignments contained insertions and/or deletions, in the scenario in Question 8?
#### Ans: `1197 + 1453 - 81 = 2569`
```
cat Arabidopsis_thaliana.local.sam | grep -v '^#' |cut -f6 | grep 'I'| wc -l
cat Arabidopsis_thaliana.local.sam | grep -v '^#' |cut -f6 | grep 'D'| wc -l
cat Arabidopsis_thaliana.local.sam | grep -v '^#' |cut -f6 | grep 'I'| grep 'D'| wc -l

```

15: How many entries were reported for Chr3?
#### Ans: `61517`
```
bcftools mpileup -Ou -f wu_0.v7.fas Arabidopsis_thaliana.sorted.bam | bcftools call -mv -Ov -o Arabidopsis_thaliana.vcf
cat Arabidopsis_thaliana.sam | grep -v "^#" | grep -v "^@" | cut -f3 | grep  "Chr3" | wc -l
```
